// @shopify/liquid-html-parser/dist/environment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Environment = exports.TagKind = void 0;
// Re-export all tag definition types so existing imports from './environment' keep working.
var tag_definitions_1 = require("./tag-definitions");
Object.defineProperty(exports, "TagKind", { enumerable: true, get: function () { return tag_definitions_1.TagKind; } });
const index_1 = require("./tags/index");
class Environment {
    builtins;
    custom;
    static _default;
    constructor(builtins = {}) {
        this.builtins = builtins;
        this.custom = new Map();
    }
    tagForName(name) {
        return this.custom.get(name) ?? this.builtins[name];
    }
    registerTag(name, definition) {
        this.custom.set(name, definition);
    }
    static default() {
        if (!Environment._default) {
            Environment._default = new Environment(index_1.builtinTags);
        }
        return Environment._default;
    }
    /** Reset singleton — for testing only */
    static resetDefault() {
        Environment._default = undefined;
    }
}
exports.Environment = Environment;
