// @shopify/liquid-html-parser/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LaxTagRecoveryError = exports.laxRecoverWhenValues = exports.laxRecoverTagMarkup = exports.laxRecoverVariable = exports.builtinTags = exports.MarkupParser = exports.MarkupTokenType = exports.tokenizeMarkup = exports.TokenType = exports.tokenize = exports.getConditionalComment = exports.BLOCKS = exports.VOID_ELEMENTS = exports.RAW_TAGS = exports.TAGS_WITHOUT_MARKUP = exports.toTolerantLiquidHtmlAST = exports.toTolerantLiquidAST = exports.findErrorNodeAtOffset = void 0;
__exportStar(require("./ast"), exports);
__exportStar(require("./types"), exports);
__exportStar(require("./errors"), exports);
var tolerant_1 = require("./tolerant");
Object.defineProperty(exports, "findErrorNodeAtOffset", { enumerable: true, get: function () { return tolerant_1.findErrorNodeAtOffset; } });
Object.defineProperty(exports, "toTolerantLiquidAST", { enumerable: true, get: function () { return tolerant_1.toTolerantLiquidAST; } });
Object.defineProperty(exports, "toTolerantLiquidHtmlAST", { enumerable: true, get: function () { return tolerant_1.toTolerantLiquidHtmlAST; } });
var grammar_1 = require("./grammar");
Object.defineProperty(exports, "TAGS_WITHOUT_MARKUP", { enumerable: true, get: function () { return grammar_1.TAGS_WITHOUT_MARKUP; } });
Object.defineProperty(exports, "RAW_TAGS", { enumerable: true, get: function () { return grammar_1.RAW_TAGS; } });
Object.defineProperty(exports, "VOID_ELEMENTS", { enumerable: true, get: function () { return grammar_1.VOID_ELEMENTS; } });
Object.defineProperty(exports, "BLOCKS", { enumerable: true, get: function () { return grammar_1.BLOCKS; } });
var conditional_comment_1 = require("./conditional-comment");
Object.defineProperty(exports, "getConditionalComment", { enumerable: true, get: function () { return conditional_comment_1.getConditionalComment; } });
var tokenizer_1 = require("./document/tokenizer");
Object.defineProperty(exports, "tokenize", { enumerable: true, get: function () { return tokenizer_1.tokenize; } });
Object.defineProperty(exports, "TokenType", { enumerable: true, get: function () { return tokenizer_1.TokenType; } });
var tokenizer_2 = require("./markup/tokenizer");
Object.defineProperty(exports, "tokenizeMarkup", { enumerable: true, get: function () { return tokenizer_2.tokenizeMarkup; } });
Object.defineProperty(exports, "MarkupTokenType", { enumerable: true, get: function () { return tokenizer_2.MarkupTokenType; } });
var parser_1 = require("./markup/parser");
Object.defineProperty(exports, "MarkupParser", { enumerable: true, get: function () { return parser_1.MarkupParser; } });
var index_1 = require("./tags/index");
Object.defineProperty(exports, "builtinTags", { enumerable: true, get: function () { return index_1.builtinTags; } });
var lax_recover_1 = require("./lax-recover");
Object.defineProperty(exports, "laxRecoverVariable", { enumerable: true, get: function () { return lax_recover_1.laxRecoverVariable; } });
Object.defineProperty(exports, "laxRecoverTagMarkup", { enumerable: true, get: function () { return lax_recover_1.laxRecoverTagMarkup; } });
Object.defineProperty(exports, "laxRecoverWhenValues", { enumerable: true, get: function () { return lax_recover_1.laxRecoverWhenValues; } });
Object.defineProperty(exports, "LaxTagRecoveryError", { enumerable: true, get: function () { return lax_recover_1.LaxTagRecoveryError; } });
