// @shopify/liquid-html-parser/dist/lax-recover.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LaxTagRecoveryError = void 0;
exports.laxRecoverVariable = laxRecoverVariable;
exports.laxRecoverTagMarkup = laxRecoverTagMarkup;
exports.laxRecoverWhenValues = laxRecoverWhenValues;
const parser_1 = require("./markup/parser");
const tokenizer_1 = require("./markup/tokenizer");
const index_1 = require("./tags/index");
const case_1 = require("./tags/case");
const tag_definitions_1 = require("./tag-definitions");
/**
 * Minimal stand-in for the `Parser` delegate. The lax-recovered tags
 * (`assign`, `echo`, `if`, `unless`, `for`) only use the `MarkupParser`
 * argument and never call back into `parseLiquidStatement`. If a future tag
 * recovery needs it, this throws loudly rather than silently mis-parsing.
 */
const noopParser = {
    parseLiquidStatement(_tagName, _markupString, _startOffset, _ctx) {
        throw new Error('lax-recover: nested statement parsing is not supported during recovery');
    },
};
/**
 * Lax-recover the `LiquidVariable` for a string-markup `{{ ... }}`.
 * Returns `undefined` when even lax parsing cannot produce a value (the caller
 * then falls back to rendering nothing / the raw string).
 */
function laxRecoverVariable(rawMarkup, source, offset) {
    try {
        const tokens = (0, tokenizer_1.tokenizeMarkup)(rawMarkup, offset);
        const parser = new parser_1.MarkupParser(tokens, source, offset, offset + rawMarkup.length).enableLax();
        const variable = parser.liquidVariable();
        return variable;
    }
    catch {
        return undefined;
    }
}
/**
 * Lax-recover the structured markup for a base-case tag (`{% name markup %}`).
 * Looks up the tag definition and re-runs its `parse` callback under lax mode.
 * The returned value is the tag-specific markup node (`AssignMarkup`,
 * `LiquidVariable`, `LiquidConditionalExpression`, `ForMarkup`, …) that the
 * strict path would have produced.
 *
 * Unlike the variable recovery, this does NOT swallow failures into an
 * empty-output sentinel. Core Liquid still surfaces errors for unrecoverable or
 * semantically-invalid tag markup even in lax mode: an unknown tag raises via
 * `block.rb`, invalid assign syntax raises via `assign.rb`, and an unsupported
 * condition operator raises `Unknown operator …` via `condition.rb`. So this
 * throws a `LaxTagRecoveryError` (for an unknown tag) or propagates the parse
 * error (for malformed markup) instead of reducing to empty output. The
 * render-tree caller translates the thrown error into the rendered error
 * surface; only genuinely-empty lax renders return normally (the tag's own
 * `parse` succeeds and produces markup).
 */
function laxRecoverTagMarkup(tagName, markupString, source, offset, markupEnd) {
    const def = index_1.builtinTags[tagName];
    if (!def || (def.kind !== tag_definitions_1.TagKind.Tag && def.kind !== tag_definitions_1.TagKind.Block)) {
        throw new LaxTagRecoveryError(`Unknown tag '${tagName}'`);
    }
    // increment/decrement lax_parse is `@variable_name = markup.strip`, which
    // accepts empty markup as the shared "" counter slot. variableLookup() throws
    // on empty input rather than producing that slot, so signal "empty markup,
    // recoverable" to the render-tree caller (which synthesizes the empty-name
    // lookup) by returning undefined here instead of letting the parse throw.
    if ((tagName === 'increment' || tagName === 'decrement') && markupString.trim() === '') {
        return undefined;
    }
    const tokens = (0, tokenizer_1.tokenizeMarkup)(markupString, offset);
    const parser = new parser_1.MarkupParser(tokens, source, offset, markupEnd).enableLax();
    return def.parse(tagName, parser, noopParser);
}
/**
 * Raised by `laxRecoverTagMarkup` when a base-case tag cannot be recovered
 * because it is not a real tag. Distinct from the parser's own
 * `LiquidHTMLASTParsingError` (thrown for malformed markup) so the render-tree
 * can recognize an unknown-tag failure specifically.
 */
class LaxTagRecoveryError extends Error {
    constructor(message) {
        super(message);
        this.name = 'LaxTagRecoveryError';
    }
}
exports.LaxTagRecoveryError = LaxTagRecoveryError;
/**
 * Lax-recover the value list for a `{% when ... %}` branch whose strict parse
 * failed (e.g. a trailing element `when 1 bar` or an invalid expression
 * `when foo=>bar`) and arrived as a raw string. `when` is a branch, not a
 * top-level tag in `builtinTags`, so it has no `TagDefinition`; its parse is
 * the standalone `whenBranchParse`, which consults lax mode to drop a trailing
 * fragment and (via the lax `variableLookup`) fold a contiguous `=>bar` into
 * the preceding lookup. Mirrors Ruby `parse_lax_when` (`case.rb:132-143`).
 * Returns `undefined` when even lax parsing fails (caller then yields no match).
 */
function laxRecoverWhenValues(markupString, source, offset, markupEnd) {
    try {
        const tokens = (0, tokenizer_1.tokenizeMarkup)(markupString, offset);
        const parser = new parser_1.MarkupParser(tokens, source, offset, markupEnd).enableLax();
        return (0, case_1.whenBranchParse)('when', parser);
    }
    catch {
        return undefined;
    }
}
