// @shopify/liquid-html-parser/dist/tolerant.js
"use strict";
/**
 * Opt-in tolerant entry points for the Liquid+HTML parser.
 *
 * These mirror `toLiquidAST` / `toLiquidHtmlAST` from `./ast` exactly, except
 * they construct a `TolerantDocumentParser` instead of the base
 * `DocumentParser`. A structural parse failure that would abort the default
 * parse is instead caught and surfaced as a `LiquidErrorNode`, so the returned
 * `DocumentNode` can carry several errors interleaved with the constructs the
 * parser did recover. The default path is a different class reached by a
 * different function and stays byte-identical.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.toTolerantLiquidAST = toTolerantLiquidAST;
exports.toTolerantLiquidHtmlAST = toTolerantLiquidHtmlAST;
exports.findErrorNodeAtOffset = findErrorNodeAtOffset;
const ast_1 = require("./ast");
const tolerant_parser_1 = require("./document/tolerant-parser");
const tokenizer_1 = require("./document/tokenizer");
const environment_1 = require("./environment");
const types_1 = require("./types");
/*
 * Tolerant variant of `toLiquidAST` (Liquid-only, `parseHtml: false`).
 * Defaults `allowUnclosedDocumentNode: true` so an unterminated document is
 * recovered rather than thrown on — the tolerant contract.
 */
function toTolerantLiquidAST(source, options = {
    allowUnclosedDocumentNode: true,
    mode: 'tolerant',
}) {
    const env = options.environment ?? environment_1.Environment.default();
    const tokens = (0, tokenizer_1.tokenize)(source);
    const parser = new tolerant_parser_1.TolerantDocumentParser(tokens, source, env, false, options.allowUnclosedDocumentNode);
    return parser.parseDocument();
}
/*
 * Tolerant variant of `toLiquidHtmlAST` (Liquid+HTML, `parseHtml: true`).
 * Defaults `allowUnclosedDocumentNode: true` so an unterminated document is
 * recovered rather than thrown on — the tolerant contract.
 */
function toTolerantLiquidHtmlAST(source, options = {
    allowUnclosedDocumentNode: true,
    mode: 'tolerant',
}) {
    const env = options.environment ?? environment_1.Environment.default();
    const tokens = (0, tokenizer_1.tokenize)(source);
    const parser = new tolerant_parser_1.TolerantDocumentParser(tokens, source, env, true, options.allowUnclosedDocumentNode);
    return parser.parseDocument();
}
/*
 * Locate the error node the caret is sitting in.
 *
 * A tolerant parse can leave several `LiquidErrorNode`s scattered through
 * the tree, one per region the parser gave up on. Completion needs the one
 * the caret is actually inside, so we return the *deepest* error node whose
 * span contains `offset` — the most specific recovery point — paired with
 * its ancestry.
 *
 * The result shape (`{ node, ancestors }`) mirrors `findCurrentNode` in the
 * language server so a later phase can swap the completion context source
 * with minimal churn. `null` means the caret is not inside any error region
 * and the existing (non-error) path should handle it.
 */
function findErrorNodeAtOffset(ast, offset) {
    /*
     * `walk` only hands each visited node its immediate parent, and it visits
     * in post-order (parents after children), so the map is only complete once
     * the traversal ends. We therefore record every parent link first, gather
     * the error nodes that contain the offset, and resolve depth/ancestry from
     * the finished map afterwards.
     */
    const parentOf = new Map();
    const candidates = [];
    (0, ast_1.walk)(ast, (node, parent) => {
        parentOf.set(node, parent);
        if (node.type === types_1.NodeTypes.LiquidErrorNode &&
            offset >= node.position.start &&
            offset <= node.position.end) {
            candidates.push(node);
        }
    });
    if (candidates.length === 0) {
        return null;
    }
    /* Ancestry, root→parent, by climbing the parent map from the node up. */
    const ancestorsOf = (node) => {
        const chain = [];
        let cursor = parentOf.get(node);
        while (cursor !== undefined) {
            chain.push(cursor);
            cursor = parentOf.get(cursor);
        }
        return chain.reverse();
    };
    /*
     * Deepest wins: when error spans nest, the innermost node has the longest
     * ancestor chain. First candidate wins ties, which is stable given walk's
     * deterministic traversal order.
     */
    let best = candidates[0];
    let bestDepth = ancestorsOf(best).length;
    for (let i = 1; i < candidates.length; i++) {
        const depth = ancestorsOf(candidates[i]).length;
        if (depth > bestDepth) {
            best = candidates[i];
            bestDepth = depth;
        }
    }
    return { node: best, ancestors: ancestorsOf(best) };
}
