// @shopify/liquid-html-parser/dist/document/factories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeDocumentNode = makeDocumentNode;
exports.envelopeFromTokens = envelopeFromTokens;
exports.makeLiquidTagBaseCase = makeLiquidTagBaseCase;
exports.makeLiquidTagNamed = makeLiquidTagNamed;
exports.makeLiquidBranchUnnamed = makeLiquidBranchUnnamed;
exports.makeLiquidBranchNamed = makeLiquidBranchNamed;
exports.makeLiquidRawTag = makeLiquidRawTag;
exports.makeLiquidVariableOutput = makeLiquidVariableOutput;
exports.makeTextNode = makeTextNode;
exports.makeRawMarkup = makeRawMarkup;
exports.makeYamlFrontmatter = makeYamlFrontmatter;
exports.makeHtmlElement = makeHtmlElement;
exports.makeHtmlVoidElement = makeHtmlVoidElement;
exports.makeHtmlSelfClosingElement = makeHtmlSelfClosingElement;
exports.makeHtmlRawNode = makeHtmlRawNode;
exports.makeHtmlComment = makeHtmlComment;
exports.makeHtmlDoctype = makeHtmlDoctype;
exports.makeAttrDoubleQuoted = makeAttrDoubleQuoted;
exports.makeAttrSingleQuoted = makeAttrSingleQuoted;
exports.makeAttrUnquoted = makeAttrUnquoted;
exports.makeAttrEmpty = makeAttrEmpty;
exports.makeHtmlDanglingMarkerClose = makeHtmlDanglingMarkerClose;
const types_1 = require("../types");
function makeDocumentNode(children, source) {
    return {
        type: types_1.NodeTypes.Document,
        children,
        name: '#document',
        _source: source,
        position: { start: 0, end: source.length },
        source,
    };
}
function envelopeFromTokens(openToken, closeToken, source) {
    const inner = source.slice(openToken.end, closeToken.start);
    const trimmed = inner.trimStart();
    const leadingSpaces = inner.length - trimmed.length;
    const firstWhitespace = trimmed.search(/\s/);
    const tagName = firstWhitespace === -1 ? trimmed.trim() : trimmed.slice(0, firstWhitespace);
    const tagNameStartInInner = leadingSpaces;
    const afterTagName = tagNameStartInInner + tagName.length;
    const markupString = inner.slice(afterTagName);
    // markupOffset points to the first non-whitespace character of the markup in source
    let markupOffsetInInner = afterTagName;
    while (markupOffsetInInner < inner.length &&
        (inner[markupOffsetInInner] === ' ' || inner[markupOffsetInInner] === '\t')) {
        markupOffsetInInner++;
    }
    const markupOffset = openToken.end + markupOffsetInInner;
    const whitespaceStart = openToken.end - openToken.start > 2 ? '-' : '';
    const whitespaceEnd = closeToken.end - closeToken.start > 2 ? '-' : '';
    return {
        tagName,
        markupString,
        markupOffset,
        markupEnd: closeToken.start,
        whitespaceStart,
        whitespaceEnd,
        blockStartPosition: { start: openToken.start, end: closeToken.end },
        source,
    };
}
function makeLiquidTagBaseCase(envelope, children, blockEndPosition, delimiterWhitespace, reason, 
// `#`-comment lines in `{% liquid %}` keep their inner indentation verbatim;
// every other base-case tag trims (the default).
preserveMarkup = false) {
    const posEnd = blockEndPosition ? blockEndPosition.end : envelope.blockStartPosition.end;
    return {
        type: types_1.NodeTypes.LiquidTag,
        name: envelope.tagName,
        markup: preserveMarkup ? envelope.markupString : envelope.markupString.trim(),
        children,
        whitespaceStart: envelope.whitespaceStart,
        whitespaceEnd: envelope.whitespaceEnd,
        delimiterWhitespaceStart: delimiterWhitespace?.start,
        delimiterWhitespaceEnd: delimiterWhitespace?.end,
        blockStartPosition: envelope.blockStartPosition,
        blockEndPosition,
        markupPosition: { start: envelope.markupOffset, end: envelope.markupEnd },
        position: { start: envelope.blockStartPosition.start, end: posEnd },
        source: envelope.source,
        reason: reason,
    };
}
function makeLiquidTagNamed(envelope, markup, children, blockEndPosition, delimiterWhitespace) {
    const posEnd = blockEndPosition ? blockEndPosition.end : envelope.blockStartPosition.end;
    return {
        type: types_1.NodeTypes.LiquidTag,
        name: envelope.tagName,
        markup,
        children,
        whitespaceStart: envelope.whitespaceStart,
        whitespaceEnd: envelope.whitespaceEnd,
        delimiterWhitespaceStart: delimiterWhitespace?.start,
        delimiterWhitespaceEnd: delimiterWhitespace?.end,
        blockStartPosition: envelope.blockStartPosition,
        blockEndPosition,
        markupPosition: { start: envelope.markupOffset, end: envelope.markupEnd },
        position: { start: envelope.blockStartPosition.start, end: posEnd },
        source: envelope.source,
    };
}
function makeLiquidBranchUnnamed(bodyStart, source) {
    return {
        type: types_1.NodeTypes.LiquidBranch,
        name: null,
        markup: '',
        children: [],
        whitespaceStart: '',
        whitespaceEnd: '',
        markupPosition: { start: bodyStart, end: bodyStart },
        blockStartPosition: { start: bodyStart, end: bodyStart },
        blockEndPosition: { start: -1, end: -1 },
        position: { start: bodyStart, end: bodyStart },
        source,
    };
}
function makeLiquidBranchNamed(envelope, markup) {
    return {
        type: types_1.NodeTypes.LiquidBranch,
        name: envelope.tagName,
        markup,
        children: [],
        whitespaceStart: envelope.whitespaceStart,
        whitespaceEnd: envelope.whitespaceEnd,
        markupPosition: { start: envelope.markupOffset, end: envelope.markupEnd },
        blockStartPosition: envelope.blockStartPosition,
        blockEndPosition: { start: -1, end: -1 },
        position: {
            start: envelope.blockStartPosition.start,
            end: envelope.blockStartPosition.end,
        },
        source: envelope.source,
    };
}
function makeLiquidRawTag(envelope, body, blockEndPosition, delimiterWhitespace) {
    let markupEnd = envelope.markupEnd;
    while (markupEnd > envelope.markupOffset &&
        (envelope.source[markupEnd - 1] === ' ' || envelope.source[markupEnd - 1] === '\t')) {
        markupEnd--;
    }
    return {
        type: types_1.NodeTypes.LiquidRawTag,
        name: envelope.tagName,
        markup: envelope.markupString.trim(),
        markupPosition: { start: envelope.markupOffset, end: markupEnd },
        body,
        whitespaceStart: envelope.whitespaceStart,
        whitespaceEnd: envelope.whitespaceEnd,
        delimiterWhitespaceStart: delimiterWhitespace.start,
        delimiterWhitespaceEnd: delimiterWhitespace.end,
        blockStartPosition: envelope.blockStartPosition,
        blockEndPosition,
        position: { start: envelope.blockStartPosition.start, end: blockEndPosition.end },
        source: envelope.source,
    };
}
function makeLiquidVariableOutput(openToken, closeToken, markup, source) {
    return {
        type: types_1.NodeTypes.LiquidVariableOutput,
        markup,
        whitespaceStart: openToken.end - openToken.start > 2 ? '-' : '',
        whitespaceEnd: closeToken.end - closeToken.start > 2 ? '-' : '',
        markupPosition: { start: openToken.end, end: closeToken.start },
        position: { start: openToken.start, end: closeToken.end },
        source,
    };
}
function makeTextNode(value, start, end, source) {
    return {
        type: types_1.NodeTypes.TextNode,
        value,
        position: { start, end },
        source,
    };
}
function makeRawMarkup(kind, value, nodes, start, end, source) {
    return {
        type: types_1.NodeTypes.RawMarkup,
        kind,
        value,
        nodes,
        position: { start, end },
        source,
    };
}
function makeYamlFrontmatter(body, start, end, source) {
    return {
        type: types_1.NodeTypes.YAMLFrontmatter,
        body,
        position: { start, end },
        source,
    };
}
function makeHtmlElement(name, attributes, children, blockStartPosition, blockEndPosition, source) {
    return {
        type: types_1.NodeTypes.HtmlElement,
        name,
        attributes,
        children,
        blockStartPosition,
        blockEndPosition,
        position: { start: blockStartPosition.start, end: blockEndPosition.end },
        source,
    };
}
function makeHtmlVoidElement(name, attributes, blockStartPosition, source) {
    return {
        type: types_1.NodeTypes.HtmlVoidElement,
        name,
        attributes,
        blockStartPosition,
        position: blockStartPosition,
        source,
    };
}
function makeHtmlSelfClosingElement(name, attributes, blockStartPosition, source) {
    return {
        type: types_1.NodeTypes.HtmlSelfClosingElement,
        name,
        attributes,
        blockStartPosition,
        position: blockStartPosition,
        source,
    };
}
function makeHtmlRawNode(name, attributes, body, blockStartPosition, blockEndPosition, source) {
    return {
        type: types_1.NodeTypes.HtmlRawNode,
        name,
        attributes,
        body,
        blockStartPosition,
        blockEndPosition,
        position: { start: blockStartPosition.start, end: blockEndPosition.end },
        source,
    };
}
function makeHtmlComment(body, start, end, source) {
    return {
        type: types_1.NodeTypes.HtmlComment,
        body,
        position: { start, end },
        source,
    };
}
function makeHtmlDoctype(legacyDoctypeString, start, end, source) {
    return {
        type: types_1.NodeTypes.HtmlDoctype,
        legacyDoctypeString,
        position: { start, end },
        source,
    };
}
function makeAttrDoubleQuoted(name, value, attributePosition, start, end, source) {
    return {
        type: types_1.NodeTypes.AttrDoubleQuoted,
        name,
        value,
        attributePosition,
        position: { start, end },
        source,
    };
}
function makeAttrSingleQuoted(name, value, attributePosition, start, end, source) {
    return {
        type: types_1.NodeTypes.AttrSingleQuoted,
        name,
        value,
        attributePosition,
        position: { start, end },
        source,
    };
}
function makeAttrUnquoted(name, value, attributePosition, start, end, source) {
    return {
        type: types_1.NodeTypes.AttrUnquoted,
        name,
        value,
        attributePosition,
        position: { start, end },
        source,
    };
}
function makeAttrEmpty(name, start, end, source) {
    return {
        type: types_1.NodeTypes.AttrEmpty,
        name,
        position: { start, end },
        source,
    };
}
function makeHtmlDanglingMarkerClose(name, start, end, source) {
    return {
        type: types_1.NodeTypes.HtmlDanglingMarkerClose,
        name,
        blockStartPosition: { start, end },
        position: { start, end },
        source,
    };
}
