// @shopify/liquid-html-parser/dist/document/liquid-lines.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseLiquidStatement = parseLiquidStatement;
exports.parseLineBlockTag = parseLineBlockTag;
exports.parseLineRawTag = parseLineRawTag;
exports.parseLineHybridTag = parseLineHybridTag;
exports.parseLineBlockBody = parseLineBlockBody;
exports.parseLineBranchedBody = parseLineBranchedBody;
exports.parseLineBranchMarkup = parseLineBranchMarkup;
const factories_1 = require("./factories");
const environment_1 = require("../environment");
const errors_1 = require("../errors");
const utils_1 = require("../utils");
const parser_1 = require("../markup/parser");
const tokenizer_1 = require("../markup/tokenizer");
const if_1 = require("../tags/if");
const case_1 = require("../tags/case");
const shared_1 = require("../shared");
const liquid_blocks_1 = require("./liquid-blocks");
const tree_builder_1 = require("./tree-builder");
const liquid_raw_1 = require("./liquid-raw");
// liquidStatement := tagName markup
function parseLiquidStatement(parser, tagName, markupString, markupOffset, ctx) {
    const line = ctx.lines[ctx.index - 1];
    const envelope = (0, shared_1.envelopeFromLine)(line, parser.getSource());
    if (tagName === '#') {
        // Preserve the inline comment's inner indentation (`preserveMarkup`); only
        // the single separator space after `#` was stripped upstream.
        return (0, factories_1.makeLiquidTagBaseCase)(envelope, undefined, undefined, undefined, undefined, true);
    }
    if (tagName.startsWith('end')) {
        throw new errors_1.LiquidHTMLASTParsingError(`Unexpected end tag '${tagName}' in {% liquid %} block`, parser.getSource(), line.nameOffset, line.lineEnd);
    }
    const def = parser.lineEnv.tagForName(tagName);
    if (!def) {
        return (0, factories_1.makeLiquidTagBaseCase)(envelope);
    }
    switch (def.kind) {
        case environment_1.TagKind.Tag: {
            let reason;
            try {
                const tokens = (0, tokenizer_1.tokenizeMarkup)(markupString, markupOffset);
                const markupParser = new parser_1.MarkupParser(tokens, parser.getSource(), markupOffset, envelope.markupEnd);
                if (parser.isTolerant())
                    markupParser.enableTolerant();
                const markup = def.parse(tagName, markupParser, parser);
                if (!markupParser.isAtEnd()) {
                    return (0, factories_1.makeLiquidTagBaseCase)(envelope, undefined, undefined, undefined, 'unexpected tokens after markup');
                }
                return (0, factories_1.makeLiquidTagNamed)(envelope, markup);
            }
            catch (e) {
                reason = e instanceof Error ? e.message : 'unknown error';
                return (0, factories_1.makeLiquidTagBaseCase)(envelope, undefined, undefined, undefined, reason);
            }
        }
        case environment_1.TagKind.Block:
            return parseLineBlockTag(parser, def, envelope, markupString, markupOffset, ctx);
        case environment_1.TagKind.Raw:
            return parseLineRawTag(parser, def, envelope, ctx);
        case environment_1.TagKind.Hybrid:
            return parseLineHybridTag(parser, def, envelope, markupString, markupOffset, ctx);
        default:
            return (0, utils_1.assertNever)(def);
    }
}
// lineBlockTag := tagName markup LF lineBlockBody | lineBranchedBody "end" tagName
function parseLineBlockTag(parser, def, envelope, markupString, markupOffset, ctx) {
    let markup;
    let markupParsed = false;
    let reason;
    try {
        const tokens = (0, tokenizer_1.tokenizeMarkup)(markupString, markupOffset);
        const markupParser = new parser_1.MarkupParser(tokens, parser.getSource());
        if (parser.isTolerant())
            markupParser.enableTolerant();
        markup = def.parse(envelope.tagName, markupParser, parser);
        if (!markupParser.isAtEnd()) {
            markup = undefined;
            reason = 'unexpected tokens after markup';
        }
        else {
            markupParsed = true;
        }
    }
    catch (e) {
        markup = undefined;
        reason = e instanceof Error ? e.message : 'unknown error';
    }
    const endTagLength = 3 + envelope.tagName.length; // 'end' + tagName
    if (def.branches.length === 0) {
        const { children, endNameOffset } = parseLineBlockBody(parser, envelope.tagName, ctx);
        const endPosition = { start: endNameOffset, end: endNameOffset + endTagLength };
        if (markupParsed) {
            return (0, factories_1.makeLiquidTagNamed)(envelope, markup, children, endPosition, { start: '', end: '' });
        }
        return (0, factories_1.makeLiquidTagBaseCase)(envelope, children, endPosition, { start: '', end: '' }, reason);
    }
    const { branches, endNameOffset } = parseLineBranchedBody(parser, envelope.tagName, def, ctx);
    const endPosition = { start: endNameOffset, end: endNameOffset + endTagLength };
    if (markupParsed) {
        return (0, factories_1.makeLiquidTagNamed)(envelope, markup, branches, endPosition, { start: '', end: '' });
    }
    return (0, factories_1.makeLiquidTagBaseCase)(envelope, branches, endPosition, { start: '', end: '' }, reason);
}
// lineRawTag := tagName LF line* "end" tagName
function parseLineRawTag(parser, _def, envelope, ctx) {
    const tagName = envelope.tagName;
    const endTagName = `end${tagName}`;
    // lineEnd is the exclusive end of the line content (one past the last
    // non-whitespace character), which is the position of the `\n` itself.
    // The RawMarkup position starts at lineEnd, but the body string starts
    // one character later (skipping the `\n`) so the value text matches the
    // actual body content.
    const positionStartOffset = ctx.index > 0 && ctx.index <= ctx.lines.length
        ? ctx.lines[ctx.index - 1].lineEnd
        : envelope.blockStartPosition.end;
    const bodyContentStart = ctx.index > 0 && ctx.index <= ctx.lines.length
        ? ctx.lines[ctx.index - 1].lineEnd + 1
        : envelope.blockStartPosition.end;
    // `comment`/`doc` bodies balance nested opens of the same tag before
    // matching their end line, mirroring the document-path scan in
    // liquid-raw.ts (Ruby comment.rb v5.13.0 `comment_tag_depth`). A nested
    // `raw` block is carved out: `raw` is first-match and does not nest, so an
    // `endcomment`/`comment` word sitting on a line inside a raw body must not
    // affect the depth. Every other raw tag (`raw`, `javascript`, `schema`,
    // `style`) keeps the original first-match scan and stays byte-identical.
    const balanced = tagName === 'comment' || tagName === 'doc';
    let endLineIndex = -1;
    if (balanced) {
        let depth = 0;
        for (let i = ctx.index; i < ctx.lines.length; i++) {
            const name = ctx.lines[i].tagName;
            if (name === 'raw') {
                // Skip past the nested raw block; its body lines never affect depth.
                i++;
                while (i < ctx.lines.length && ctx.lines[i].tagName !== 'endraw')
                    i++;
                continue;
            }
            if (name === tagName) {
                depth++;
            }
            else if (name === endTagName) {
                if (depth === 0) {
                    endLineIndex = i;
                    break;
                }
                depth--;
            }
        }
    }
    else {
        for (let i = ctx.index; i < ctx.lines.length; i++) {
            if (ctx.lines[i].tagName === endTagName) {
                endLineIndex = i;
                break;
            }
        }
    }
    if (endLineIndex === -1) {
        throw new errors_1.LiquidHTMLASTParsingError(`Unclosed raw tag '${tagName}' in {% liquid %} block`, parser.getSource(), envelope.blockStartPosition.start, envelope.blockStartPosition.end);
    }
    const endLine = ctx.lines[endLineIndex];
    const bodyEndOffset = endLine.nameOffset;
    const source = parser.getSource();
    const bodyString = source.slice(bodyContentStart, bodyEndOffset);
    // Build body nodes: a TextNode wrapping the body content, matching
    // the non-liquid path in parseRawTagBody (liquid-raw.ts).
    const bodyNodes = bodyContentStart < bodyEndOffset
        ? (0, tree_builder_1.filterChildren)(tree_builder_1.ChildFilterMode.Syntactic, [(0, factories_1.makeTextNode)(bodyString, bodyContentStart, bodyEndOffset, source)], source)
        : [];
    const body = (0, factories_1.makeRawMarkup)((0, liquid_raw_1.rawMarkupKindForTag)(tagName, bodyString), bodyString, bodyNodes, positionStartOffset, bodyEndOffset, source);
    const endPosition = {
        start: endLine.nameOffset,
        end: endLine.nameOffset + endTagName.length,
    };
    ctx.index = endLineIndex + 1;
    return (0, factories_1.makeLiquidRawTag)(envelope, body, endPosition, { start: '', end: '' });
}
// lineHybridTag := tagName markup (LF lineBlockBody "end" tagName)?
function parseLineHybridTag(parser, def, envelope, markupString, markupOffset, ctx) {
    let markup;
    let markupParsed = false;
    let reason;
    try {
        const tokens = (0, tokenizer_1.tokenizeMarkup)(markupString, markupOffset);
        const markupParser = new parser_1.MarkupParser(tokens, parser.getSource());
        if (parser.isTolerant())
            markupParser.enableTolerant();
        markup = def.parse(envelope.tagName, markupParser, parser);
        if (!markupParser.isAtEnd()) {
            markup = undefined;
            reason = 'unexpected tokens after markup';
        }
        else {
            markupParsed = true;
        }
    }
    catch (e) {
        markup = undefined;
        reason = e instanceof Error ? e.message : 'unknown error';
    }
    const endTagName = `end${envelope.tagName}`;
    let hasEndTag = false;
    let depth = 0;
    for (let i = ctx.index; i < ctx.lines.length; i++) {
        if (ctx.lines[i].tagName === envelope.tagName) {
            depth++;
        }
        else if (ctx.lines[i].tagName === endTagName) {
            if (depth === 0) {
                hasEndTag = true;
                break;
            }
            depth--;
        }
    }
    if (!hasEndTag) {
        if (markupParsed) {
            return (0, factories_1.makeLiquidTagNamed)(envelope, markup);
        }
        return (0, factories_1.makeLiquidTagBaseCase)(envelope, undefined, undefined, undefined, reason);
    }
    const { children, endNameOffset } = parseLineBlockBody(parser, envelope.tagName, ctx);
    const endTagLength = 3 + envelope.tagName.length; // 'end' + tagName
    const endPosition = { start: endNameOffset, end: endNameOffset + endTagLength };
    if (markupParsed) {
        return (0, factories_1.makeLiquidTagNamed)(envelope, markup, children, endPosition, { start: '', end: '' });
    }
    return (0, factories_1.makeLiquidTagBaseCase)(envelope, children, endPosition, { start: '', end: '' }, reason);
}
// lineBlockBody := liquidStatement* "end" tagName
function parseLineBlockBody(parser, parentName, ctx) {
    const children = [];
    while (ctx.index < ctx.lines.length) {
        const line = ctx.lines[ctx.index];
        if (line.tagName === `end${parentName}`) {
            ctx.index++;
            return { children, endNameOffset: line.nameOffset };
        }
        ctx.index++;
        children.push(parseLiquidStatement(parser, line.tagName, line.markup, line.markupOffset, ctx));
    }
    if (parser.lineAllowUnclosedDocumentNode) {
        const endNameOffset = ctx.lines.length > 0 ? ctx.lines[ctx.lines.length - 1].lineEnd : 0;
        return { children, endNameOffset };
    }
    throw new errors_1.LiquidHTMLASTParsingError(`Unclosed block tag '${parentName}' in {% liquid %} block`, parser.getSource(), 0, 0);
}
// lineBranchedBody := lineBranch+ "end" tagName
function parseLineBranchedBody(parser, parentName, def, ctx) {
    const branches = [];
    const bodyStart = ctx.index > 0 ? ctx.lines[ctx.index - 1].lineEnd : 0;
    let currentBranch = (0, factories_1.makeLiquidBranchUnnamed)(bodyStart, parser.getSource());
    let currentChildren = [];
    while (ctx.index < ctx.lines.length) {
        const line = ctx.lines[ctx.index];
        if (line.tagName === `end${parentName}`) {
            (0, liquid_blocks_1.finalizeBranch)(currentBranch, currentChildren, line.nameOffset, parser.getSource(), parser.lineParseHtml ? tree_builder_1.ChildFilterMode.StripEdges : tree_builder_1.ChildFilterMode.Syntactic);
            branches.push(currentBranch);
            ctx.index++;
            return { branches, endNameOffset: line.nameOffset };
        }
        if ((0, liquid_blocks_1.isBranchName)(line.tagName) && def.branches.includes(line.tagName)) {
            (0, liquid_blocks_1.finalizeBranch)(currentBranch, currentChildren, line.nameOffset, parser.getSource(), parser.lineParseHtml ? tree_builder_1.ChildFilterMode.StripEdges : tree_builder_1.ChildFilterMode.Syntactic);
            branches.push(currentBranch);
            const branchEnvelope = (0, shared_1.envelopeFromLine)(line, parser.getSource());
            const branchMarkup = parseLineBranchMarkup(line.tagName, branchEnvelope, parser.getSource(), parser.isTolerant());
            currentBranch = (0, factories_1.makeLiquidBranchNamed)(branchEnvelope, branchMarkup);
            currentChildren = [];
            ctx.index++;
            continue;
        }
        ctx.index++;
        currentChildren.push(parseLiquidStatement(parser, line.tagName, line.markup, line.markupOffset, ctx));
    }
    if (parser.lineAllowUnclosedDocumentNode) {
        const endNameOffset = ctx.lines.length > 0 ? ctx.lines[ctx.lines.length - 1].lineEnd : 0;
        (0, liquid_blocks_1.finalizeBranch)(currentBranch, currentChildren, endNameOffset, parser.getSource(), parser.lineParseHtml ? tree_builder_1.ChildFilterMode.StripEdges : tree_builder_1.ChildFilterMode.Syntactic);
        branches.push(currentBranch);
        return { branches, endNameOffset };
    }
    throw new errors_1.LiquidHTMLASTParsingError(`Unclosed block tag '${parentName}' in {% liquid %} block`, parser.getSource(), 0, 0);
}
function parseLineBranchMarkup(branchName, envelope, source, tolerant = false) {
    switch (branchName) {
        case 'elsif': {
            try {
                const tokens = (0, tokenizer_1.tokenizeMarkup)(envelope.markupString, envelope.markupOffset);
                const markupParser = new parser_1.MarkupParser(tokens, source);
                if (tolerant)
                    markupParser.enableTolerant();
                const result = (0, if_1.elsifBranchParse)(branchName, markupParser);
                if (!markupParser.isAtEnd())
                    return envelope.markupString.trim();
                return result;
            }
            catch {
                return envelope.markupString.trim();
            }
        }
        case 'when': {
            try {
                const tokens = (0, tokenizer_1.tokenizeMarkup)(envelope.markupString, envelope.markupOffset);
                const markupParser = new parser_1.MarkupParser(tokens, source);
                if (tolerant)
                    markupParser.enableTolerant();
                const result = (0, case_1.whenBranchParse)(branchName, markupParser);
                if (!markupParser.isAtEnd())
                    return envelope.markupString.trim();
                return result;
            }
            catch {
                return envelope.markupString.trim();
            }
        }
        case 'else':
            return '';
        default:
            return (0, utils_1.assertNever)(branchName);
    }
}
