// @shopify/liquid-html-parser/dist/document/liquid-blocks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseBlockTag = parseBlockTag;
exports.parseBlockBody = parseBlockBody;
exports.parseBranchedBody = parseBranchedBody;
exports.finalizeBranch = finalizeBranch;
exports.parseBranchMarkup = parseBranchMarkup;
exports.peekTagName = peekTagName;
exports.isBlockTerminator = isBlockTerminator;
exports.consumeEndTag = consumeEndTag;
exports.isBranchName = isBranchName;
const tokenizer_1 = require("./tokenizer");
const factories_1 = require("./factories");
const tree_builder_1 = require("./tree-builder");
const errors_1 = require("../errors");
const utils_1 = require("../utils");
const parser_1 = require("../markup/parser");
const tokenizer_2 = require("../markup/tokenizer");
const if_1 = require("../tags/if");
const case_1 = require("../tags/case");
// blockTag := "{%" tagName markup "%}" branchedBody | blockBody "{%" "end" tagName "%}"
function parseBlockTag(parser, def, envelope, closeToken) {
    let markup;
    let markupParsed = false;
    let reason;
    try {
        const markupStringStart = closeToken.start - envelope.markupString.length;
        const tokens = (0, tokenizer_2.tokenizeMarkup)(envelope.markupString, markupStringStart);
        const markupParser = new parser_1.MarkupParser(tokens, parser.getSource());
        if (parser.isTolerant())
            markupParser.enableTolerant();
        markup = def.parse(envelope.tagName, markupParser, parser);
        if (!markupParser.isAtEnd()) {
            markup = undefined;
            reason = 'unexpected tokens after markup';
        }
        else {
            markupParsed = true;
        }
    }
    catch (e) {
        markup = undefined;
        reason = e instanceof Error ? e.message : 'unknown error';
    }
    const savedAllowUnclosed = parser.blockAllowUnclosedHtml;
    parser.blockAllowUnclosedHtml = CONDITIONAL_TAGS.has(envelope.tagName);
    try {
        if (def.branches.length === 0) {
            const { children, endPosition, endWhitespace } = parseBlockBody(parser, envelope);
            if (markupParsed) {
                return (0, factories_1.makeLiquidTagNamed)(envelope, markup, children, endPosition, endWhitespace);
            }
            return (0, factories_1.makeLiquidTagBaseCase)(envelope, children, endPosition, endWhitespace, reason);
        }
        const { branches, endPosition, endWhitespace } = parseBranchedBody(parser, envelope, def);
        if (markupParsed) {
            return (0, factories_1.makeLiquidTagNamed)(envelope, markup, branches, endPosition, endWhitespace);
        }
        return (0, factories_1.makeLiquidTagBaseCase)(envelope, branches, endPosition, endWhitespace, reason);
    }
    finally {
        parser.blockAllowUnclosedHtml = savedAllowUnclosed;
    }
}
// blockBody := node* "{%" "end" tagName "%}"
function parseBlockBody(parser, envelope) {
    const parentName = envelope.tagName;
    const children = [];
    while (!parser.isAtEnd()) {
        if (parser.check(tokenizer_1.TokenType.LiquidTagOpen)) {
            const tagName = parser.peekTagName();
            if (tagName === `end${parentName}`) {
                const end = parser.consumeEndTag();
                const merged = (0, tree_builder_1.filterChildren)(tree_builder_1.ChildFilterMode.Syntactic, children, parser.getSource());
                return { children: merged, endPosition: end.position, endWhitespace: end.whitespace };
            }
            if (tagName && tagName.startsWith('end')) {
                const innerName = tagName.slice(3);
                if (parser.blockEnv.tagForName(innerName)) {
                    throw new errors_1.LiquidHTMLASTParsingError(`Attempting to close LiquidTag '${innerName}' before LiquidTag '${parentName}' was closed`, parser.getSource(), envelope.blockStartPosition.start, envelope.blockStartPosition.end);
                }
            }
            if (tagName && (tagName === 'else' || tagName === 'elsif' || tagName === 'when')) {
                throw new errors_1.LiquidHTMLASTParsingError(`Attempting to open LiquidBranch '${tagName}' before LiquidTag '${parentName}' was closed`, parser.getSource(), envelope.blockStartPosition.start, envelope.blockStartPosition.end);
            }
        }
        const progressStart = parser.getPosition();
        if (parser.blockInAttributeContext && !parser.blockInAttributeValueContext) {
            children.push(...parser.parseBranchAttributes());
        }
        else {
            children.push(parser.parseNode());
        }
        if (parser.getPosition() === progressStart) {
            if (!parser.blockAllowUnclosedDocumentNode) {
                throwUnexpectedBlockTag(parser, parentName);
            }
            // Tolerant mode: the current token stalled the loop — an orphan tag the block can
            // neither absorb nor close (e.g. `{% endwhen %}` naming a branch keyword). Consume it
            // as a degraded node so the cursor advances and the loop can still reach the real
            // terminator (e.g. `{% endcapture %}`), instead of leaking the orphan up to the outer
            // parseAttributeList where a sibling close tag would hit the unguarded throw in
            // parseLiquidTag.
            if (parser.check(tokenizer_1.TokenType.LiquidTagOpen)) {
                children.push(parser.parseLiquidTag());
                continue;
            }
            break;
        }
    }
    {
        const merged = (0, tree_builder_1.filterChildren)(tree_builder_1.ChildFilterMode.Syntactic, children, parser.getSource());
        const endPos = parser.peek().start;
        const blockEndPosition = { start: endPos, end: endPos };
        if (parser.blockAllowUnclosedDocumentNode) {
            return {
                children: merged,
                endPosition: blockEndPosition,
                endWhitespace: { start: '', end: '' },
            };
        }
        throw new errors_1.LiquidHTMLASTParsingError(`Attempting to end parsing before LiquidTag '${parentName}' was closed`, parser.getSource(), envelope.blockStartPosition.start, envelope.blockStartPosition.end);
    }
}
// branchedBody := branch+ "{%" "end" tagName "%}"
function parseBranchedBody(parser, envelope, def) {
    const parentName = envelope.tagName;
    const bodyStart = envelope.blockStartPosition.end;
    const branches = [];
    let currentBranch = (0, factories_1.makeLiquidBranchUnnamed)(bodyStart, parser.getSource());
    let currentChildren = [];
    const childFilterMode = parser.blockInAttributeValueContext
        ? tree_builder_1.ChildFilterMode.Preserve
        : tree_builder_1.ChildFilterMode.Syntactic;
    while (!parser.isAtEnd()) {
        if (parser.check(tokenizer_1.TokenType.LiquidTagOpen)) {
            const tagName = parser.peekTagName();
            if (tagName === `end${parentName}`) {
                finalizeBranch(currentBranch, currentChildren, parser.peek().start, parser.getSource(), childFilterMode);
                branches.push(currentBranch);
                const end = parser.consumeEndTag();
                return { branches, endPosition: end.position, endWhitespace: end.whitespace };
            }
            if (isBranchName(tagName) && def.branches.includes(tagName)) {
                finalizeBranch(currentBranch, currentChildren, parser.peek().start, parser.getSource(), childFilterMode);
                branches.push(currentBranch);
                const openToken = parser.consume(tokenizer_1.TokenType.LiquidTagOpen);
                parser.accept(tokenizer_1.TokenType.Text);
                const closeToken = parser.consume(tokenizer_1.TokenType.LiquidTagClose);
                const branchEnvelope = (0, factories_1.envelopeFromTokens)(openToken, closeToken, parser.getSource());
                const branchMarkup = parseBranchMarkup(parser, tagName, branchEnvelope);
                currentBranch = (0, factories_1.makeLiquidBranchNamed)(branchEnvelope, branchMarkup);
                currentChildren = [];
                continue;
            }
            if (tagName && tagName.startsWith('end')) {
                const innerName = tagName.slice(3);
                if (parser.blockEnv.tagForName(innerName)) {
                    throw new errors_1.LiquidHTMLASTParsingError(`Attempting to close LiquidTag '${innerName}' before LiquidTag '${parentName}' was closed`, parser.getSource(), envelope.blockStartPosition.start, envelope.blockStartPosition.end);
                }
            }
        }
        const progressStart = parser.getPosition();
        if (parser.blockInAttributeContext && !parser.blockInAttributeValueContext) {
            currentChildren.push(...parser.parseBranchAttributes());
        }
        else {
            currentChildren.push(parser.parseNode());
        }
        if (parser.getPosition() === progressStart) {
            if (!parser.blockAllowUnclosedDocumentNode) {
                throwUnexpectedBlockTag(parser, parentName);
            }
            // Tolerant mode: the current token stalled the loop — an orphan tag the block can
            // neither absorb nor close (e.g. `{% endwhen %}` naming a branch keyword). Consume it
            // as a degraded node so the cursor advances and the loop can still reach the real
            // terminator (e.g. `{% endcase %}`), instead of leaking the orphan up to the outer
            // parseAttributeList where a sibling close tag would hit the unguarded throw in
            // parseLiquidTag.
            if (parser.check(tokenizer_1.TokenType.LiquidTagOpen)) {
                currentChildren.push(parser.parseLiquidTag());
                continue;
            }
            break;
        }
    }
    {
        const endPos = parser.peek().start;
        finalizeBranch(currentBranch, currentChildren, endPos, parser.getSource(), childFilterMode);
        branches.push(currentBranch);
        const blockEndPosition = { start: endPos, end: endPos };
        if (parser.blockAllowUnclosedDocumentNode) {
            return {
                branches,
                endPosition: blockEndPosition,
                endWhitespace: { start: '', end: '' },
            };
        }
        throw new errors_1.LiquidHTMLASTParsingError(`Attempting to end parsing before LiquidTag '${parentName}' was closed`, parser.getSource(), envelope.blockStartPosition.start, envelope.blockStartPosition.end);
    }
}
function finalizeBranch(branch, children, endPos, source, mode = tree_builder_1.ChildFilterMode.StripEdges) {
    const allMerged = (0, tree_builder_1.mergeAdjacentTextNodes)(children, source);
    branch.children = (0, tree_builder_1.filterChildren)(mode, children, source);
    branch.blockEndPosition = { start: endPos, end: endPos };
    branch.position.end = endPos;
    if (allMerged.length > 0) {
        const lastChildEnd = allMerged[allMerged.length - 1].position.end;
        if (lastChildEnd > endPos) {
            branch.position.end = lastChildEnd;
        }
    }
}
// branchMarkup := elsifMarkup | whenMarkup | ""
function parseBranchMarkup(parser, branchName, envelope) {
    const closeTokenWidth = envelope.whitespaceEnd === '-' ? 3 : 2;
    const closeTokenStart = envelope.blockStartPosition.end - closeTokenWidth;
    const markupStringStart = closeTokenStart - envelope.markupString.length;
    switch (branchName) {
        case 'elsif': {
            try {
                const tokens = (0, tokenizer_2.tokenizeMarkup)(envelope.markupString, markupStringStart);
                const markupParser = new parser_1.MarkupParser(tokens, parser.getSource());
                if (parser.isTolerant())
                    markupParser.enableTolerant();
                const result = (0, if_1.elsifBranchParse)(branchName, markupParser);
                if (!markupParser.isAtEnd())
                    return envelope.markupString.trim();
                return result;
            }
            catch {
                return envelope.markupString.trim();
            }
        }
        case 'when': {
            try {
                const tokens = (0, tokenizer_2.tokenizeMarkup)(envelope.markupString, markupStringStart);
                const markupParser = new parser_1.MarkupParser(tokens, parser.getSource());
                if (parser.isTolerant())
                    markupParser.enableTolerant();
                const result = (0, case_1.whenBranchParse)(branchName, markupParser);
                if (!markupParser.isAtEnd())
                    return envelope.markupString.trim();
                return result;
            }
            catch {
                return envelope.markupString.trim();
            }
        }
        case 'else':
            return '';
        default:
            return (0, utils_1.assertNever)(branchName);
    }
}
function peekTagName(parser) {
    if (!parser.check(tokenizer_1.TokenType.LiquidTagOpen))
        return null;
    const inner = parser.check(tokenizer_1.TokenType.Text, 1)
        ? parser
            .getSource()
            .slice(parser.tokenAt(parser.getPosition() + 1).start, parser.tokenAt(parser.getPosition() + 1).end)
        : '';
    const trimmed = inner.trimStart();
    const firstWs = trimmed.search(/\s/);
    return firstWs === -1 ? trimmed.trim() : trimmed.slice(0, firstWs);
}
function isBlockTerminator(parser) {
    if (!parser.check(tokenizer_1.TokenType.LiquidTagOpen))
        return false;
    const tagName = peekTagName(parser);
    if (!tagName)
        return false;
    return (tagName.startsWith('end') || tagName === 'elsif' || tagName === 'else' || tagName === 'when');
}
/**
 * A block-body iteration that consumed no tokens means the current token is one
 * the block can neither absorb nor close (e.g. an orphan `{% endwhen %}` naming a
 * branch keyword rather than a registered tag); looping again would spin forever.
 * In strict mode this throws a clean parse error; tolerant callers should break
 * and let the graceful "unclosed" return degrade the result.
 */
function throwUnexpectedBlockTag(parser, parentName) {
    const tok = parser.peek();
    const tagName = peekTagName(parser);
    throw new errors_1.LiquidHTMLASTParsingError(`Unexpected Liquid tag '${tagName ?? tok.type}' inside LiquidTag '${parentName}'`, parser.getSource(), tok.start, tok.end);
}
function consumeEndTag(parser) {
    const openToken = parser.consume(tokenizer_1.TokenType.LiquidTagOpen);
    parser.accept(tokenizer_1.TokenType.Text);
    const closeToken = parser.consume(tokenizer_1.TokenType.LiquidTagClose);
    const wsStart = openToken.end - openToken.start > 2 ? '-' : '';
    const wsEnd = closeToken.end - closeToken.start > 2 ? '-' : '';
    return {
        position: { start: openToken.start, end: closeToken.end },
        whitespace: { start: wsStart, end: wsEnd },
    };
}
function isBranchName(value) {
    return value === 'elsif' || value === 'else' || value === 'when';
}
const CONDITIONAL_TAGS = new Set(['if', 'unless', 'case']);
