// @shopify/liquid-html-parser/dist/document/parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.finalizeBranch = exports.DocumentParser = void 0;
const tokenizer_1 = require("./tokenizer");
const base_1 = require("./base");
const factories_1 = require("./factories");
const tree_builder_1 = require("./tree-builder");
require("../environment"); // side-effect: ensures builtinTags init before downstream imports
const liquid_variable_output_1 = require("./liquid-variable-output");
const liquid_blocks_1 = require("./liquid-blocks");
const liquid_raw_1 = require("./liquid-raw");
const liquid_tags_1 = require("./liquid-tags");
const html_1 = require("./html");
const liquid_lines_1 = require("./liquid-lines");
const node_dispatch_1 = require("./node-dispatch");
class DocumentParser extends base_1.ParserBase {
    env;
    _parseHtml;
    _allowUnclosedHtml;
    _allowUnclosedDocumentNode;
    _inAttributeContext = false;
    _inAttributeValueContext = false;
    constructor(tokens, source, env, parseHtml, allowUnclosedDocumentNode = false) {
        super(tokens, source);
        this.env = env;
        this._parseHtml = parseHtml;
        this._allowUnclosedHtml = allowUnclosedDocumentNode;
        this._allowUnclosedDocumentNode = allowUnclosedDocumentNode;
    }
    get rawParseHtml() {
        return this._parseHtml;
    }
    set rawParseHtml(v) {
        this._parseHtml = v;
    }
    get blockEnv() {
        return this.env;
    }
    get blockParseHtml() {
        return this._parseHtml;
    }
    get blockAllowUnclosedHtml() {
        return this._allowUnclosedHtml;
    }
    set blockAllowUnclosedHtml(v) {
        this._allowUnclosedHtml = v;
    }
    get blockAllowUnclosedDocumentNode() {
        return this._allowUnclosedDocumentNode;
    }
    get blockInAttributeContext() {
        return this._inAttributeContext;
    }
    get blockInAttributeValueContext() {
        return this._inAttributeValueContext;
    }
    set blockInAttributeValueContext(v) {
        this._inAttributeValueContext = v;
    }
    get htmlParseHtml() {
        return this._parseHtml;
    }
    get htmlAllowUnclosedHtml() {
        return this._allowUnclosedHtml;
    }
    set htmlAllowUnclosedHtml(v) {
        this._allowUnclosedHtml = v;
    }
    get htmlInAttributeContext() {
        return this._inAttributeContext;
    }
    set htmlInAttributeContext(v) {
        this._inAttributeContext = v;
    }
    get htmlInAttributeValueContext() {
        return this._inAttributeValueContext;
    }
    set htmlInAttributeValueContext(v) {
        this._inAttributeValueContext = v;
    }
    get lineEnv() {
        return this.env;
    }
    get lineParseHtml() {
        return this._parseHtml;
    }
    get lineAllowUnclosedDocumentNode() {
        return this._allowUnclosedDocumentNode;
    }
    get dispatchParseHtml() {
        return this._parseHtml;
    }
    get dispatchAllowUnclosedHtml() {
        return this._allowUnclosedHtml;
    }
    tagForName(name) {
        return this.env.tagForName(name);
    }
    // document := yamlFrontmatter? node*
    parseDocument() {
        const children = [];
        if (this.check(tokenizer_1.TokenType.YamlFrontmatter))
            children.push((0, node_dispatch_1.parseNode)(this));
        while (!this.isAtEnd())
            children.push(this.parseNode());
        return (0, factories_1.makeDocumentNode)((0, tree_builder_1.filterChildren)(tree_builder_1.ChildFilterMode.Syntactic, children, this.source), this.source);
    }
    parseNode() {
        return (0, node_dispatch_1.parseNode)(this);
    }
    parseLiquidTag() {
        return (0, liquid_tags_1.parseLiquidTag)(this);
    }
    parseLiquidVariableOutput() {
        return (0, liquid_variable_output_1.parseLiquidVariableOutput)(this);
    }
    peekTagName() {
        return (0, liquid_blocks_1.peekTagName)(this);
    }
    isBlockTerminator() {
        return (0, liquid_blocks_1.isBlockTerminator)(this);
    }
    consumeEndTag() {
        return (0, liquid_blocks_1.consumeEndTag)(this);
    }
    parseLiquidInRange(bodyStart, bodyEnd) {
        return (0, liquid_raw_1.parseLiquidInRange)(this, bodyStart, bodyEnd);
    }
    parseBranchAttributes() {
        return (0, html_1.parseBranchAttributesImpl)(this);
    }
    parseLiquidStatement(tagName, markupString, markupOffset, ctx) {
        return (0, liquid_lines_1.parseLiquidStatement)(this, tagName, markupString, markupOffset, ctx);
    }
    parseHtmlElement() {
        return (0, html_1.parseHtmlElement)(this);
    }
    parseHtmlComment() {
        return (0, html_1.parseHtmlComment)(this);
    }
    parseHtmlDoctype() {
        return (0, html_1.parseHtmlDoctype)(this);
    }
    parseOrphanedHtmlCloseTag() {
        return (0, html_1.parseOrphanedHtmlCloseTag)(this);
    }
    parseHtmlDanglingMarkerClose() {
        return (0, html_1.parseHtmlDanglingMarkerClose)(this);
    }
}
exports.DocumentParser = DocumentParser;
var liquid_blocks_2 = require("./liquid-blocks");
Object.defineProperty(exports, "finalizeBranch", { enumerable: true, get: function () { return liquid_blocks_2.finalizeBranch; } });
