// @shopify/liquid-html-parser/dist/document/tokenizer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenType = void 0;
exports.tokenize = tokenize;
const utils_1 = require("../utils");
var TokenType;
(function (TokenType) {
    TokenType["Text"] = "Text";
    TokenType["LiquidTagOpen"] = "LiquidTagOpen";
    TokenType["LiquidTagClose"] = "LiquidTagClose";
    TokenType["LiquidVariableOutputOpen"] = "LiquidVariableOutputOpen";
    TokenType["LiquidVariableOutputClose"] = "LiquidVariableOutputClose";
    TokenType["HtmlTagOpen"] = "HtmlTagOpen";
    TokenType["HtmlCloseTagOpen"] = "HtmlCloseTagOpen";
    TokenType["HtmlTagClose"] = "HtmlTagClose";
    TokenType["HtmlSelfClose"] = "HtmlSelfClose";
    TokenType["HtmlCommentOpen"] = "HtmlCommentOpen";
    TokenType["HtmlCommentClose"] = "HtmlCommentClose";
    TokenType["HtmlDoctypeOpen"] = "HtmlDoctypeOpen";
    TokenType["HtmlEquals"] = "HtmlEquals";
    TokenType["HtmlQuoteOpen"] = "HtmlQuoteOpen";
    TokenType["HtmlQuoteClose"] = "HtmlQuoteClose";
    TokenType["YamlFrontmatter"] = "YamlFrontmatter";
    TokenType["EndOfInput"] = "EndOfInput";
})(TokenType || (exports.TokenType = TokenType = {}));
function tokenize(source, options = {}) {
    const tokens = [];
    const modeStack = [];
    let mode = Mode.Default;
    let pos = 0;
    let textStart = -1;
    let quoteChar = '';
    // Resume inside a quoted attribute value when reslicing a suffix that begins
    // mid-attribute (e.g. `...{% endraw %}">` straddled by a stray `{{`). The
    // enclosing HtmlTag mode is kept beneath QuotedValue so popping on the
    // closing quote correctly emits the attribute close and trailing `>`.
    if (options.insideQuotedAttribute) {
        quoteChar = options.insideQuotedAttribute;
        modeStack.push(Mode.HtmlTag);
        mode = Mode.QuotedValue;
    }
    else if (options.insideHtmlTag) {
        // Resume inside the attribute list of an open tag. The empty mode stack
        // means the closing `>` pops back to `Default`, so the rest of the suffix
        // (`>ok</div>`) tokenizes as normal document content.
        mode = Mode.HtmlTag;
    }
    function ch(offset) {
        const i = pos + offset;
        return i < source.length ? source[i] : '';
    }
    function match(s) {
        return source.startsWith(s, pos);
    }
    function flushText() {
        if (textStart !== -1) {
            tokens.push({ type: TokenType.Text, start: textStart, end: pos });
            textStart = -1;
        }
    }
    function emit(type, length) {
        flushText();
        tokens.push({ type, start: pos, end: pos + length });
        pos += length;
    }
    function pushMode(next) {
        modeStack.push(mode);
        mode = next;
    }
    function popMode() {
        mode = modeStack.length > 0 ? modeStack.pop() : Mode.Default;
    }
    function startText() {
        if (textStart === -1)
            textStart = pos;
    }
    // YAML frontmatter: only at position 0
    if (!options.skipFrontmatter && (match('---\n') || match('---\r\n'))) {
        const searchStart = source.indexOf('\n', 0) + 1;
        const closeIdx = source.indexOf('\n---', searchStart);
        if (closeIdx !== -1) {
            let end = closeIdx + 4; // after \n---
            // Include trailing newline/CRLF
            if (end < source.length && source[end] === '\r')
                end++;
            if (end < source.length && source[end] === '\n')
                end++;
            tokens.push({ type: TokenType.YamlFrontmatter, start: 0, end });
            pos = end;
        }
    }
    // Liquid open check — reused in Default, HtmlTag, and QuotedValue modes
    function scanLiquidOpen() {
        if (match('{{-')) {
            emit(TokenType.LiquidVariableOutputOpen, 3);
            pushMode(Mode.LiquidVariableOutput);
            return true;
        }
        if (match('{{')) {
            emit(TokenType.LiquidVariableOutputOpen, 2);
            pushMode(Mode.LiquidVariableOutput);
            return true;
        }
        if (match('{%-')) {
            emit(TokenType.LiquidTagOpen, 3);
            pushMode(Mode.LiquidTag);
            return true;
        }
        if (match('{%')) {
            emit(TokenType.LiquidTagOpen, 2);
            pushMode(Mode.LiquidTag);
            return true;
        }
        return false;
    }
    while (pos < source.length) {
        switch (mode) {
            case Mode.LiquidTag: {
                if (match('-%}')) {
                    emit(TokenType.LiquidTagClose, 3);
                    popMode();
                }
                else if (match('%}')) {
                    emit(TokenType.LiquidTagClose, 2);
                    popMode();
                }
                else {
                    startText();
                    pos++;
                }
                break;
            }
            case Mode.LiquidVariableOutput: {
                if (match('-}}')) {
                    emit(TokenType.LiquidVariableOutputClose, 3);
                    popMode();
                }
                else if (match('}}')) {
                    emit(TokenType.LiquidVariableOutputClose, 2);
                    popMode();
                }
                else {
                    startText();
                    pos++;
                }
                break;
            }
            case Mode.Default: {
                if (scanLiquidOpen())
                    continue;
                if (match('<!--')) {
                    emit(TokenType.HtmlCommentOpen, 4);
                    continue;
                }
                if (match('-->')) {
                    emit(TokenType.HtmlCommentClose, 3);
                    continue;
                }
                if (match('<!')) {
                    emit(TokenType.HtmlDoctypeOpen, 2);
                    pushMode(Mode.HtmlTag);
                    continue;
                }
                if (match('</')) {
                    const after = ch(2);
                    if (/[a-zA-Z]/.test(after) || after === '{') {
                        emit(TokenType.HtmlCloseTagOpen, 2);
                        pushMode(Mode.HtmlTag);
                        continue;
                    }
                }
                if (ch(0) === '<') {
                    const after = ch(1);
                    if (/[a-zA-Z]/.test(after) || after === '{') {
                        emit(TokenType.HtmlTagOpen, 1);
                        pushMode(Mode.HtmlTag);
                        continue;
                    }
                }
                startText();
                pos++;
                break;
            }
            case Mode.HtmlTag: {
                if (scanLiquidOpen())
                    continue;
                if (match('/>')) {
                    emit(TokenType.HtmlSelfClose, 2);
                    popMode();
                    continue;
                }
                if (ch(0) === '>') {
                    emit(TokenType.HtmlTagClose, 1);
                    popMode();
                    continue;
                }
                if (ch(0) === '=') {
                    emit(TokenType.HtmlEquals, 1);
                    continue;
                }
                // Accept straight quotes and curly (smart) quotes as attribute-value
                // openers. Curly quotes get normalized to straight quotes downstream;
                // recognizing them here (HTML scope only — Liquid tokenization is
                // untouched) stops the value from splitting at interior spaces.
                if (ch(0) === '"' ||
                    ch(0) === "'" ||
                    ch(0) === '“' ||
                    ch(0) === '”' ||
                    ch(0) === '‘' ||
                    ch(0) === '’') {
                    quoteChar = ch(0);
                    emit(TokenType.HtmlQuoteOpen, 1);
                    pushMode(Mode.QuotedValue);
                    continue;
                }
                startText();
                pos++;
                break;
            }
            case Mode.QuotedValue: {
                if (scanLiquidOpen())
                    continue;
                // Curly quotes are directional, so a value opened with a left curly
                // quote closes on its right partner (and vice-versa); straight quotes
                // close on themselves.
                if (ch(0) === closingQuoteFor(quoteChar) || ch(0) === quoteChar) {
                    emit(TokenType.HtmlQuoteClose, 1);
                    popMode();
                    continue;
                }
                startText();
                pos++;
                break;
            }
            default:
                (0, utils_1.assertNever)(mode);
        }
    }
    flushText();
    tokens.push({ type: TokenType.EndOfInput, start: source.length, end: source.length });
    return tokens;
}
var Mode;
(function (Mode) {
    Mode["Default"] = "Default";
    Mode["HtmlTag"] = "HtmlTag";
    Mode["QuotedValue"] = "QuotedValue";
    Mode["LiquidTag"] = "LiquidTag";
    Mode["LiquidVariableOutput"] = "LiquidVariableOutput";
})(Mode || (Mode = {}));
// Curly (smart) quotes come in directional pairs: a value opened with a left
// curly quote closes on its right partner and vice-versa. Straight quotes are
// their own partner, so they close on an identical character.
function closingQuoteFor(open) {
    switch (open) {
        case '“':
            return '”';
        case '”':
            return '“';
        case '‘':
            return '’';
        case '’':
            return '‘';
        default:
            return open;
    }
}
