// @shopify/liquid-html-parser/dist/liquid-doc/factories.js
"use strict";
/**
 * Factory functions for LiquidDoc AST nodes.
 *
 * Follows the same pattern as document/factories.ts — pure constructors,
 * no parsing logic.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeTextNode = makeTextNode;
exports.makeLiquidDocParamNode = makeLiquidDocParamNode;
exports.makeLiquidDocDescriptionNode = makeLiquidDocDescriptionNode;
exports.makeLiquidDocExampleNode = makeLiquidDocExampleNode;
exports.makeLiquidDocPromptNode = makeLiquidDocPromptNode;
const types_1 = require("../types");
function makeTextNode(value, start, end, source) {
    return {
        type: types_1.NodeTypes.TextNode,
        value,
        position: { start, end },
        source,
    };
}
function makeLiquidDocParamNode(paramName, paramType, paramDescription, required, start, end, source) {
    return {
        type: types_1.NodeTypes.LiquidDocParamNode,
        name: 'param',
        paramName,
        paramType,
        paramDescription,
        required,
        position: { start, end },
        source,
    };
}
function makeLiquidDocDescriptionNode(content, isImplicit, isInline, start, end, source) {
    return {
        type: types_1.NodeTypes.LiquidDocDescriptionNode,
        name: 'description',
        content,
        isImplicit,
        isInline,
        position: { start, end },
        source,
    };
}
function makeLiquidDocExampleNode(content, isInline, start, end, source) {
    return {
        type: types_1.NodeTypes.LiquidDocExampleNode,
        name: 'example',
        content,
        isInline,
        position: { start, end },
        source,
    };
}
function makeLiquidDocPromptNode(content, start, end, source) {
    return {
        type: types_1.NodeTypes.LiquidDocPromptNode,
        name: 'prompt',
        content,
        position: { start, end },
        source,
    };
}
