// @shopify/liquid-html-parser/dist/liquid-doc/parser.js
"use strict";
/**
 * Recursive-descent parser for LiquidDoc body content.
 *
 * Consumes level-1 tokens (Annotation, TextLine, Newline) from the body
 * tokenizer. For @param, re-tokenizes the inline content using the level-2
 * param tokenizer to extract {type}, [name], -, and description.
 *
 * No raw string scanning in this file — all extraction is token-based.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LiquidDocParser = void 0;
exports.parseLiquidDoc = parseLiquidDoc;
const tokenizer_1 = require("./tokenizer");
const tokenizer_2 = require("./tokenizer");
const factories_1 = require("./factories");
const utils_1 = require("../utils");
/**
 * Parse a LiquidDoc body string into AST nodes.
 */
function parseLiquidDoc(body, startOffset, source) {
    const tokens = (0, tokenizer_2.tokenizeLiquidDoc)(body, startOffset);
    const parser = new LiquidDocParser(tokens, source);
    return parser.parse();
}
class LiquidDocParser {
    tokens;
    source;
    p;
    constructor(tokens, source) {
        this.tokens = tokens;
        this.source = source;
        this.p = 0;
    }
    /**
     * Parse the full LiquidDoc body into a list of nodes.
     */
    // liquidDoc := implicitDescription? annotation*
    parse() {
        const nodes = [];
        this.skipBlankLines();
        // Implicit description: text before any @annotation
        if (this.check(tokenizer_1.LiquidDocTokenType.TextLine)) {
            nodes.push(this.parseImplicitDescription());
        }
        this.skipBlankLines();
        while (!this.isAtEnd()) {
            if (this.check(tokenizer_1.LiquidDocTokenType.Newline)) {
                this.skipNewlines();
                continue;
            }
            if (this.check(tokenizer_1.LiquidDocTokenType.Annotation)) {
                nodes.push(this.parseAnnotation());
                continue;
            }
            // Skip whitespace-only text lines in the main loop
            if (this.check(tokenizer_1.LiquidDocTokenType.TextLine) && this.peek().value.trim() === '') {
                this.advance();
                continue;
            }
            // Unexpected text — collect consecutive text lines into a single TextNode
            if (this.check(tokenizer_1.LiquidDocTokenType.TextLine)) {
                const { value, startPos, endPos } = this.collectContentLines({
                    trimTrailingNewlines: true,
                });
                if (value.trim().length > 0) {
                    nodes.push((0, factories_1.makeTextNode)(value, startPos, endPos, this.source));
                }
                continue;
            }
            // Any other unexpected token — consume as single text fallback
            const tok = this.advance();
            nodes.push((0, factories_1.makeTextNode)(tok.value, tok.start, tok.end, this.source));
        }
        return nodes;
    }
    // implicitDescription := textLine+
    parseImplicitDescription() {
        const { value, startPos, endPos } = this.collectContentLines();
        const content = (0, factories_1.makeTextNode)(value, startPos, endPos, this.source);
        return (0, factories_1.makeLiquidDocDescriptionNode)(content, true, true, startPos, endPos, this.source);
    }
    // annotation := "@" annotationName content
    parseAnnotation() {
        const annotationToken = this.consume(tokenizer_1.LiquidDocTokenType.Annotation);
        const name = annotationToken.value;
        // `@descriptionText` — no space between `@description` and its content. The
        // level-1 tokenizer greedily folds the leading text into the annotation
        // name (`descriptionText`), so recognize the `description` prefix here and
        // treat the glued remainder as the start of the inline description content.
        // This matches the pre-swap parser, which printed `@description Text`.
        if (name !== LiquidDocAnnotation.Description &&
            name.startsWith(LiquidDocAnnotation.Description)) {
            return this.parseGluedDescription(annotationToken);
        }
        if (!isKnownAnnotation(name)) {
            return this.parseUnsupportedAnnotation(annotationToken);
        }
        switch (name) {
            case LiquidDocAnnotation.Param:
                return this.parseParam(annotationToken);
            case LiquidDocAnnotation.Description:
                return this.parseDescription(annotationToken);
            case LiquidDocAnnotation.Example:
                return this.parseExample(annotationToken);
            case LiquidDocAnnotation.Prompt:
                return this.parsePrompt(annotationToken);
            default:
                (0, utils_1.assertNever)(name);
        }
    }
    /**
     * Parse @param using the level-2 param tokenizer.
     *
     * The first TextLine after `@param` is re-tokenized into {type}, [name], -, word, text.
     * Subsequent lines (collected via collectContentLines) are appended to the description.
     */
    // param := "@param" ("{" type "}")? ("[" name "]" | name) ("-" description)?
    parseParam(annotationToken) {
        const start = annotationToken.start;
        // Get the inline content on the same line as @param
        const inlineToken = this.accept(tokenizer_1.LiquidDocTokenType.TextLine);
        if (!inlineToken) {
            // @param with nothing after it
            const emptyName = (0, factories_1.makeTextNode)('', annotationToken.end, annotationToken.end, this.source);
            return (0, factories_1.makeLiquidDocParamNode)(emptyName, null, null, true, start, annotationToken.end, this.source);
        }
        // Re-tokenize the inline content for param-specific grammar
        const walker = new ParamTokenWalker((0, tokenizer_2.tokenizeParamContent)(inlineToken.value, inlineToken.start));
        // Optional type: {type}
        let paramType = null;
        const typeToken = walker.accept(tokenizer_1.ParamTokenType.Type);
        if (typeToken) {
            paramType = (0, factories_1.makeTextNode)(typeToken.value, typeToken.start + 1, typeToken.end - 1, this.source);
        }
        // Name: [name] (optional) or Word (required)
        let paramName;
        let required;
        const optNameToken = walker.accept(tokenizer_1.ParamTokenType.OptionalName);
        if (optNameToken) {
            required = false;
            paramName = (0, factories_1.makeTextNode)(optNameToken.value, optNameToken.start + 1, optNameToken.end - 1, this.source);
        }
        else {
            required = true;
            const wordToken = walker.accept(tokenizer_1.ParamTokenType.Word);
            if (wordToken) {
                paramName = (0, factories_1.makeTextNode)(wordToken.value, wordToken.start, wordToken.end, this.source);
            }
            else {
                paramName = (0, factories_1.makeTextNode)('', annotationToken.end, annotationToken.end, this.source);
            }
        }
        // Optional dash + description
        let paramDescription = null;
        const dashToken = walker.accept(tokenizer_1.ParamTokenType.Dash);
        // Collect remaining param tokens as inline description (source-slice to preserve whitespace)
        let inlineDescStart = -1;
        let inlineDescEnd = -1;
        while (!walker.isAtEnd()) {
            const t = walker.advance();
            if (inlineDescStart === -1)
                inlineDescStart = t.start;
            inlineDescEnd = t.end;
        }
        const inlineDesc = inlineDescStart !== -1 ? this.source.slice(inlineDescStart, inlineDescEnd) : '';
        // Collect multiline continuation (stop at paragraph breaks for @param, trim trailing newlines)
        const { value: continuation, startPos: _contStart, endPos: contEnd, } = this.collectContentLines({ stopAtParagraphBreak: true, trimTrailingNewlines: true });
        const hasContinuation = continuation.trim().length > 0;
        if (inlineDesc || hasContinuation) {
            const fullDesc = hasContinuation ? inlineDesc + continuation : inlineDesc;
            const descStart = inlineDescStart !== -1
                ? inlineDescStart
                : dashToken
                    ? dashToken.end + 1
                    : annotationToken.end;
            const descEnd = hasContinuation ? contEnd : inlineDescEnd;
            paramDescription = (0, factories_1.makeTextNode)(fullDesc, descStart, descEnd, this.source);
        }
        const end = paramDescription?.position.end ?? inlineToken.end;
        return (0, factories_1.makeLiquidDocParamNode)(paramName, paramType, paramDescription, required, start, end, this.source);
    }
    // description := "@description" content
    parseDescription(annotationToken) {
        const start = annotationToken.start;
        const inlineToken = this.accept(tokenizer_1.LiquidDocTokenType.TextLine);
        const isInline = inlineToken !== null;
        if (isInline) {
            const { value: rest, endPos } = this.collectContentLines();
            const hasCont = rest.trim().length > 0;
            const fullContent = hasCont ? inlineToken.value + rest : inlineToken.value;
            const end = hasCont ? endPos : inlineToken.end;
            const content = (0, factories_1.makeTextNode)(fullContent, inlineToken.start, end, this.source);
            return (0, factories_1.makeLiquidDocDescriptionNode)(content, false, true, start, end, this.source);
        }
        this.skipBlankLines();
        const { value, startPos, endPos } = this.collectContentLines();
        const contentStart = value ? startPos : annotationToken.end;
        const content = (0, factories_1.makeTextNode)(value || '', contentStart, value ? endPos : contentStart, this.source);
        return (0, factories_1.makeLiquidDocDescriptionNode)(content, false, false, start, value ? endPos : contentStart, this.source);
    }
    // gluedDescription := "@description" gluedContent  (no separating space)
    //
    // The glued suffix and any same-line text that follows are one inline
    // description. Content is read straight from source (not by concatenating
    // token values) so the original spacing between the suffix and the trailing
    // text is preserved — the suffix's own token boundary is mid-word.
    parseGluedDescription(annotationToken) {
        const start = annotationToken.start;
        // Source offset immediately past the literal `@description`; the glued
        // remainder (e.g. `This` in `@descriptionThis`) begins here.
        const contentStart = start + 1 + LiquidDocAnnotation.Description.length;
        const inlineToken = this.accept(tokenizer_1.LiquidDocTokenType.TextLine);
        const { value: rest, endPos } = this.collectContentLines();
        const hasRest = rest.trim().length > 0;
        let end;
        if (hasRest)
            end = endPos;
        else if (inlineToken)
            end = inlineToken.end;
        else
            end = annotationToken.end;
        const content = (0, factories_1.makeTextNode)(this.source.slice(contentStart, end), contentStart, end, this.source);
        return (0, factories_1.makeLiquidDocDescriptionNode)(content, false, true, start, end, this.source);
    }
    // example := "@example" content
    parseExample(annotationToken) {
        const start = annotationToken.start;
        const inlineToken = this.accept(tokenizer_1.LiquidDocTokenType.TextLine);
        const isInline = inlineToken !== null;
        if (isInline) {
            const { value: rest, endPos } = this.collectContentLines();
            const hasCont = rest.trim().length > 0;
            const fullContent = hasCont ? inlineToken.value + rest : inlineToken.value;
            const end = hasCont ? endPos : inlineToken.end;
            const content = (0, factories_1.makeTextNode)(fullContent, inlineToken.start, end, this.source);
            return (0, factories_1.makeLiquidDocExampleNode)(content, true, start, end, this.source);
        }
        // Non-inline: skip initial blank lines before the example content
        this.skipBlankLines();
        const { value, startPos, endPos } = this.collectContentLines();
        const contentStart = value ? startPos : annotationToken.end;
        const content = (0, factories_1.makeTextNode)(value || '', contentStart, value ? endPos : contentStart, this.source);
        return (0, factories_1.makeLiquidDocExampleNode)(content, false, start, value ? endPos : contentStart, this.source);
    }
    // prompt := "@prompt" content
    parsePrompt(annotationToken) {
        const start = annotationToken.start;
        const { value, startPos, endPos } = this.collectContentLines();
        const contentStart = value ? startPos : annotationToken.end;
        const content = (0, factories_1.makeTextNode)(value || '', contentStart, value ? endPos : contentStart, this.source);
        return (0, factories_1.makeLiquidDocPromptNode)(content, start, value ? endPos : contentStart, this.source);
    }
    // unsupportedAnnotation := "@" name content
    parseUnsupportedAnnotation(annotationToken) {
        const start = annotationToken.start;
        const inlineToken = this.accept(tokenizer_1.LiquidDocTokenType.TextLine);
        if (inlineToken) {
            const fullValue = this.source.slice(start, inlineToken.end);
            return (0, factories_1.makeTextNode)(fullValue, start, inlineToken.end, this.source);
        }
        const fullValue = '@' + annotationToken.value;
        return (0, factories_1.makeTextNode)(fullValue, start, annotationToken.end, this.source);
    }
    peek() {
        return this.tokens[this.p];
    }
    consume(type) {
        const token = this.peek();
        if (token.type !== type) {
            throw new Error(`Expected ${type} but found ${token.type} (${JSON.stringify(token.value)})`);
        }
        this.advance();
        return token;
    }
    accept(type) {
        if (this.peek().type !== type)
            return null;
        return this.advance();
    }
    check(type) {
        return this.tokens[this.p].type === type;
    }
    isAtEnd() {
        return this.check(tokenizer_1.LiquidDocTokenType.EndOfInput);
    }
    /** Consume the current token unconditionally and advance. */
    advance() {
        return this.tokens[this.p++];
    }
    skipNewlines() {
        let skipped = false;
        while (this.check(tokenizer_1.LiquidDocTokenType.Newline)) {
            this.advance();
            skipped = true;
        }
        return skipped;
    }
    /** Skip newlines and whitespace-only text lines (e.g. indentation with no content). */
    skipBlankLines() {
        while (!this.isAtEnd()) {
            if (this.check(tokenizer_1.LiquidDocTokenType.Newline)) {
                this.advance();
                continue;
            }
            if (this.check(tokenizer_1.LiquidDocTokenType.TextLine) && this.peek().value.trim() === '') {
                this.advance();
                continue;
            }
            break;
        }
    }
    /**
     * Collect text lines and newlines until the next @annotation or end of input.
     * Returns the concatenated content and end position.
     *
     * Options:
     * - stopAtParagraphBreak: stop at blank line + non-annotation text (for @param)
     * - trimTrailingNewlines: strip trailing \n from value (for @param + free text)
     */
    collectContentLines(opts = {}) {
        const { stopAtParagraphBreak = false, trimTrailingNewlines = false } = opts;
        let value = '';
        let startPos = this.tokens[this.p]?.start ?? this.tokens[this.p - 1].end;
        let endPos = startPos;
        while (!this.isAtEnd()) {
            const current = this.peek();
            if (this.check(tokenizer_1.LiquidDocTokenType.Annotation))
                break;
            if (this.check(tokenizer_1.LiquidDocTokenType.Newline)) {
                // Look ahead past newlines and whitespace-only text to see if we should stop
                let lookAhead = this.p + 1;
                while (lookAhead < this.tokens.length) {
                    const la = this.tokens[lookAhead];
                    if (la.type === tokenizer_1.LiquidDocTokenType.Newline) {
                        lookAhead++;
                        continue;
                    }
                    if (la.type === tokenizer_1.LiquidDocTokenType.TextLine && la.value.trim() === '') {
                        lookAhead++;
                        continue;
                    }
                    break;
                }
                // Paragraph break: blank line(s) followed by non-annotation text.
                // Only applies to @param continuation — implicit descriptions span across blank lines.
                const newlineCount = lookAhead - this.p;
                if (stopAtParagraphBreak &&
                    newlineCount >= 2 &&
                    lookAhead < this.tokens.length &&
                    this.tokens[lookAhead].type === tokenizer_1.LiquidDocTokenType.TextLine &&
                    this.tokens[lookAhead].value.trim() !== '') {
                    break;
                }
                // Trailing newlines before annotation boundary.
                if (lookAhead < this.tokens.length &&
                    this.tokens[lookAhead].type === tokenizer_1.LiquidDocTokenType.Annotation) {
                    while (this.p < lookAhead &&
                        (this.check(tokenizer_1.LiquidDocTokenType.Newline) ||
                            (this.check(tokenizer_1.LiquidDocTokenType.TextLine) && this.peek().value.trim() === ''))) {
                        if (!trimTrailingNewlines) {
                            if (this.check(tokenizer_1.LiquidDocTokenType.Newline))
                                value += '\n';
                            endPos = this.peek().end;
                        }
                        this.advance();
                    }
                    break;
                }
                // Trailing newlines before end-of-input.
                if (lookAhead >= this.tokens.length ||
                    this.tokens[lookAhead].type === tokenizer_1.LiquidDocTokenType.EndOfInput) {
                    while (!this.isAtEnd() &&
                        (this.check(tokenizer_1.LiquidDocTokenType.Newline) ||
                            (this.check(tokenizer_1.LiquidDocTokenType.TextLine) && this.peek().value.trim() === ''))) {
                        if (!trimTrailingNewlines) {
                            if (this.check(tokenizer_1.LiquidDocTokenType.Newline))
                                value += '\n';
                            endPos = this.peek().end;
                        }
                        this.advance();
                    }
                    break;
                }
                value += '\n';
                endPos = current.end;
                this.advance();
                continue;
            }
            // Skip whitespace-only text lines at the boundary (before end or annotation)
            if (this.check(tokenizer_1.LiquidDocTokenType.TextLine) && current.value.trim() === '') {
                this.advance();
                continue;
            }
            // TextLine — strip leading whitespace (indentation) on first line
            if (value === '') {
                const leadingWs = current.value.length - current.value.trimStart().length;
                startPos = current.start + leadingWs;
            }
            const trimmedLine = value === '' ? current.value.trimStart() : current.value;
            value += trimmedLine;
            endPos = current.end;
            this.advance();
        }
        return { value, startPos, endPos };
    }
}
exports.LiquidDocParser = LiquidDocParser;
/** Known annotation names that get dedicated node types. */
var LiquidDocAnnotation;
(function (LiquidDocAnnotation) {
    LiquidDocAnnotation["Param"] = "param";
    LiquidDocAnnotation["Description"] = "description";
    LiquidDocAnnotation["Example"] = "example";
    LiquidDocAnnotation["Prompt"] = "prompt";
})(LiquidDocAnnotation || (LiquidDocAnnotation = {}));
function isKnownAnnotation(value) {
    return (value === LiquidDocAnnotation.Param ||
        value === LiquidDocAnnotation.Description ||
        value === LiquidDocAnnotation.Example ||
        value === LiquidDocAnnotation.Prompt);
}
class ParamTokenWalker {
    tokens;
    p = 0;
    constructor(tokens) {
        this.tokens = tokens;
    }
    accept(type) {
        if (this.tokens[this.p].type !== type)
            return null;
        return this.tokens[this.p++];
    }
    check(type) {
        return this.tokens[this.p].type === type;
    }
    isAtEnd() {
        return this.check(tokenizer_1.ParamTokenType.EndOfInput);
    }
    /** Current token (for reading value/position without consuming). */
    current() {
        return this.tokens[this.p];
    }
    /** Consume the current token unconditionally and advance. */
    advance() {
        return this.tokens[this.p++];
    }
}
