// @shopify/liquid-html-parser/dist/markup/expression-adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.COMPARISON_OPERATORS = exports.EQUALITY_OPERATORS = exports.LOGICAL_OPERATORS = exports.ComparisonOperator = exports.EqualityOperator = exports.LogicalOperator = void 0;
exports.isBinaryExpression = isBinaryExpression;
exports.adaptConditional = adaptConditional;
exports.adaptComplex = adaptComplex;
const types_1 = require("../types");
const utils_1 = require("../utils");
var LogicalOperator;
(function (LogicalOperator) {
    LogicalOperator["And"] = "and";
    LogicalOperator["Or"] = "or";
})(LogicalOperator || (exports.LogicalOperator = LogicalOperator = {}));
var EqualityOperator;
(function (EqualityOperator) {
    EqualityOperator["Equal"] = "==";
    EqualityOperator["NotEqual"] = "!=";
})(EqualityOperator || (exports.EqualityOperator = EqualityOperator = {}));
var ComparisonOperator;
(function (ComparisonOperator) {
    ComparisonOperator["LessThan"] = "<";
    ComparisonOperator["GreaterThan"] = ">";
    ComparisonOperator["LessThanOrEqual"] = "<=";
    ComparisonOperator["GreaterThanOrEqual"] = ">=";
    ComparisonOperator["Contains"] = "contains";
})(ComparisonOperator || (exports.ComparisonOperator = ComparisonOperator = {}));
exports.LOGICAL_OPERATORS = new Set(Object.values(LogicalOperator));
exports.EQUALITY_OPERATORS = new Set(Object.values(EqualityOperator));
exports.COMPARISON_OPERATORS = new Set(Object.values(ComparisonOperator));
function isBinaryExpression(node) {
    return 'kind' in node;
}
function toRelation(op) {
    switch (op) {
        case LogicalOperator.And:
            return 'and';
        case LogicalOperator.Or:
            return 'or';
        default:
            return (0, utils_1.assertNever)(op);
    }
}
function toComparators(op) {
    switch (op) {
        case EqualityOperator.Equal:
            return types_1.Comparators.EQUAL;
        case EqualityOperator.NotEqual:
            return types_1.Comparators.NOT_EQUAL;
        case ComparisonOperator.LessThan:
            return types_1.Comparators.LESS_THAN;
        case ComparisonOperator.GreaterThan:
            return types_1.Comparators.GREATER_THAN;
        case ComparisonOperator.LessThanOrEqual:
            return types_1.Comparators.LESS_THAN_OR_EQUAL;
        case ComparisonOperator.GreaterThanOrEqual:
            return types_1.Comparators.GREATER_THAN_OR_EQUAL;
        case ComparisonOperator.Contains:
            return types_1.Comparators.CONTAINS;
        default:
            // Lax recovery preserves an unknown operator (`=`, or a bare word like
            // `true`) in the comparison node's `op`. It is not a known comparator, so
            // it cannot be mapped to the `Comparators` enum here; pass the raw string
            // through unchanged so the engine raises `Unknown operator <op>` at
            // evaluate time (mirrors Ruby `condition.rb:186`). Strict parsing never
            // produces an unknown operator, so this branch is lax-condition only.
            return op;
    }
}
function adaptConditional(node) {
    if (!isBinaryExpression(node)) {
        return node;
    }
    switch (node.kind) {
        case 'logical': {
            const right = adaptConditional(node.right);
            // The original parser sets a nested LogicalExpression's position.start
            // to include the preceding operator keyword (`and`/`or`).
            if (right.type === types_1.NodeTypes.LogicalExpression) {
                right.position = { ...right.position, start: node.opStart };
            }
            return {
                type: types_1.NodeTypes.LogicalExpression,
                relation: toRelation(node.op),
                left: adaptConditional(node.left),
                right,
                position: node.position,
                source: node.source,
            };
        }
        case 'comparison':
            return {
                type: types_1.NodeTypes.Comparison,
                comparator: toComparators(node.op),
                left: node.left,
                right: node.right,
                position: node.position,
                source: node.source,
            };
        default:
            return (0, utils_1.assertNever)(node);
    }
}
function adaptComplex(node) {
    if (!isBinaryExpression(node)) {
        return node;
    }
    return {
        type: types_1.NodeTypes.BooleanExpression,
        condition: adaptConditional(node),
        position: node.position,
        source: node.source,
    };
}
