// @shopify/liquid-html-parser/dist/tags/block.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.blockTag = void 0;
const tokenizer_1 = require("../markup/tokenizer");
const types_1 = require("../types");
const tag_definitions_1 = require("../tag-definitions");
const BLOCK_TYPE_REGEX = /^_?[a-zA-Z0-9][\w-]*$/;
exports.blockTag = {
    kind: tag_definitions_1.TagKind.Block,
    branches: [],
    parse(_name, markup, _parser) {
        if (markup.isAtEnd()) {
            throw new Error("in 'block' - missing block type");
        }
        const name = markup.valueExpression();
        if (name.type !== types_1.NodeTypes.String) {
            throw new Error("in 'block' - file name must be a string literal");
        }
        if (!BLOCK_TYPE_REGEX.test(name.value)) {
            throw new Error(`in 'block' - '${name.value}' is not a valid block type`);
        }
        let args = [];
        if (markup.consumeOptional(tokenizer_1.MarkupTokenType.Comma) && !markup.isAtEnd()) {
            args = markup.blockNamedArguments();
        }
        if (!markup.isAtEnd()) {
            const token = markup.peek();
            throw new Error(`Unexpected token in 'block' tag: ${token.value}`);
        }
        return {
            type: types_1.NodeTypes.BlockMarkup,
            name,
            args,
            position: { start: name.position.start, end: markup.peek().start },
            source: name.source,
        };
    },
};
