// @shopify/liquid-html-parser/dist/tags/content-for.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.contentForTag = void 0;
const tokenizer_1 = require("../markup/tokenizer");
const types_1 = require("../types");
const tag_definitions_1 = require("../tag-definitions");
exports.contentForTag = {
    kind: tag_definitions_1.TagKind.Tag,
    parse(_name, markup, _parser) {
        const contentForType = markup.valueExpression();
        if (contentForType.type !== types_1.NodeTypes.String) {
            throw new Error(`Expected string for content_for type, got ${contentForType.type}`);
        }
        const args = [];
        markup.consumeOptional(tokenizer_1.MarkupTokenType.Comma);
        while (markup.look(tokenizer_1.MarkupTokenType.Id)) {
            try {
                args.push(markup.namedArgument());
            }
            catch {
                break;
            }
            markup.consumeOptional(tokenizer_1.MarkupTokenType.Comma);
        }
        return {
            type: types_1.NodeTypes.ContentForMarkup,
            contentForType,
            args,
            position: { start: contentForType.position.start, end: markup.peek().start },
            source: contentForType.source,
        };
    },
};
