// @shopify/liquid-html-parser/dist/tags/cycle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cycleTag = void 0;
const tokenizer_1 = require("../markup/tokenizer");
const types_1 = require("../types");
const tag_definitions_1 = require("../tag-definitions");
exports.cycleTag = {
    kind: tag_definitions_1.TagKind.Tag,
    parse(_name, markup, _parser) {
        const first = markup.valueExpression();
        let groupName = null;
        const args = [];
        if (markup.consumeOptional(tokenizer_1.MarkupTokenType.Colon)) {
            groupName = first;
            args.push(markup.valueExpression());
        }
        else {
            args.push(first);
        }
        while (markup.consumeOptional(tokenizer_1.MarkupTokenType.Comma)) {
            // Lax: tolerate a trailing comma (`{% cycle "1", "2", %}`). Ruby's
            // `variables_from_string` splits on commas and `.compact`s away the empty
            // trailing segment, and `strict2_parse` does `break if end_of_string`.
            // We only relax this on the lax-recovery path so strict `toLiquidHtmlAST`
            // output (consumed by the theme-check linter) is unchanged (DD-7): a
            // trailing comma stays a strict syntax error there.
            if (markup.isLax() && markup.isAtEnd())
                break;
            args.push(markup.valueExpression());
        }
        return {
            type: types_1.NodeTypes.CycleMarkup,
            groupName,
            args,
            position: { start: first.position.start, end: markup.peek().start },
            source: first.source,
        };
    },
};
