// @shopify/liquid-html-parser/dist/tags/for.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tablerowTag = exports.forTag = void 0;
const tokenizer_1 = require("../markup/tokenizer");
const types_1 = require("../types");
const tag_definitions_1 = require("../tag-definitions");
function parseForMarkup(_name, markup, _parser) {
    const nameToken = markup.consume(tokenizer_1.MarkupTokenType.Id);
    if (!markup.id('in')) {
        throw new Error("Expected 'in'");
    }
    const collection = markup.valueExpression();
    const reversed = markup.id('reversed');
    // Named args in for/tablerow accept optional commas between args
    // (matching Ruby's `p.consume?(:comma)` behavior).
    const args = [];
    while (!markup.isAtEnd()) {
        markup.consumeOptional(tokenizer_1.MarkupTokenType.Comma);
        if (markup.isLax()) {
            // Lax recovery (render-tree only; `toLiquidHtmlAST` parses strict, so this
            // branch never runs there): Ruby's lax `tablerow`/`for` parse stops at the
            // first unparseable named argument and discards the rest rather than
            // erroring — e.g. `{% tablerow i in (1..5) limit: foo=>bar %}` keeps
            // `limit: foo` and drops `=>bar`. `namedArgument()` always consumes at
            // least its leading id or throws, so a throw is a clean stop that leaves
            // the already-parsed args intact (no infinite loop).
            try {
                args.push(markup.namedArgument());
            }
            catch {
                break;
            }
        }
        else {
            args.push(markup.namedArgument());
        }
    }
    return {
        type: types_1.NodeTypes.ForMarkup,
        variableName: nameToken.value,
        collection,
        reversed,
        args,
        position: { start: nameToken.start, end: markup.peek().start },
        source: collection.source,
    };
}
exports.forTag = {
    kind: tag_definitions_1.TagKind.Block,
    branches: ['else'],
    parse: parseForMarkup,
};
exports.tablerowTag = {
    kind: tag_definitions_1.TagKind.Block,
    branches: ['else'],
    parse: parseForMarkup,
};
