// @shopify/liquid-html-parser/dist/tags/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.builtinTags = void 0;
const if_1 = require("./if");
const case_1 = require("./case");
const for_1 = require("./for");
const assign_1 = require("./assign");
const echo_1 = require("./echo");
const render_1 = require("./render");
const cycle_1 = require("./cycle");
const section_1 = require("./section");
const content_for_1 = require("./content-for");
const block_1 = require("./block");
const paginate_1 = require("./paginate");
const form_1 = require("./form");
const increment_1 = require("./increment");
const decrement_1 = require("./decrement");
const capture_1 = require("./capture");
const layout_1 = require("./layout");
const sections_1 = require("./sections");
const partial_1 = require("./partial");
const ifchanged_1 = require("./ifchanged");
const break_1 = require("./break");
const continue_1 = require("./continue");
const raw_1 = require("./raw");
const liquid_1 = require("./liquid");
exports.builtinTags = {
    assign: assign_1.assignTag,
    block: block_1.blockTag,
    break: break_1.breakTag,
    capture: capture_1.captureTag,
    case: case_1.caseTag,
    content_for: content_for_1.contentForTag,
    continue: continue_1.continueTag,
    cycle: cycle_1.cycleTag,
    decrement: decrement_1.decrementTag,
    echo: echo_1.echoTag,
    for: for_1.forTag,
    form: form_1.formTag,
    if: if_1.ifTag,
    ifchanged: ifchanged_1.ifchangedTag,
    include: render_1.includeTag,
    increment: increment_1.incrementTag,
    layout: layout_1.layoutTag,
    liquid: liquid_1.liquidTag,
    paginate: paginate_1.paginateTag,
    partial: partial_1.partialTag,
    render: render_1.renderTag,
    section: section_1.sectionTag,
    sections: sections_1.sectionsTag,
    tablerow: for_1.tablerowTag,
    unless: if_1.unlessTag,
    comment: raw_1.commentRaw,
    doc: raw_1.docRaw,
    javascript: raw_1.javascriptRaw,
    raw: raw_1.rawRaw,
    schema: raw_1.schemaRaw,
    style: raw_1.styleRaw,
    stylesheet: raw_1.stylesheetRaw,
};
