// @shopify/liquid-html-parser/dist/tags/paginate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.paginateTag = void 0;
const tokenizer_1 = require("../markup/tokenizer");
const types_1 = require("../types");
const tag_definitions_1 = require("../tag-definitions");
exports.paginateTag = {
    kind: tag_definitions_1.TagKind.Block,
    branches: [],
    parse(_name, markup, _parser) {
        const collection = markup.valueExpression();
        if (!markup.id('by')) {
            throw new Error("Expected 'by' in paginate tag");
        }
        const pageSize = markup.valueExpression();
        const args = [];
        if (markup.consumeOptional(tokenizer_1.MarkupTokenType.Comma)) {
            args.push(...markup.namedArguments());
        }
        else if (markup.isTolerant() && markup.peek().type === tokenizer_1.MarkupTokenType.Id) {
            // Tolerant: paginate accepts whitespace-separated named attrs with no
            // leading comma (`... by N window_size: 50`). Consume them so the tag
            // builds as PaginateMarkup instead of leaving leftover tokens. Strict is
            // unchanged.
            args.push(...markup.namedArguments());
        }
        return {
            type: types_1.NodeTypes.PaginateMarkup,
            collection,
            pageSize,
            args,
            position: { start: collection.position.start, end: markup.peek().start },
            source: collection.source,
        };
    },
};
