// @shopify/liquid-html-parser/dist/tags/raw.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stylesheetRaw = exports.styleRaw = exports.schemaRaw = exports.javascriptRaw = exports.rawRaw = exports.docRaw = exports.commentRaw = void 0;
const tag_definitions_1 = require("../tag-definitions");
function parseNoMarkup(name, markup, _parser) {
    if (!markup.isAtEnd()) {
        throw new Error(`'${name}' tag does not accept markup`);
    }
}
exports.commentRaw = {
    kind: tag_definitions_1.TagKind.Raw,
    parseLiquidInBody: false,
    parse: parseNoMarkup,
};
// doc bodies are parsed separately by the liquid-doc-parser — parseLiquidInBody
// is intentionally omitted (undefined) so the parser can detect this and hand
// the body off to the doc-specific parser rather than the standard raw pipeline.
exports.docRaw = {
    kind: tag_definitions_1.TagKind.Raw,
    parse: parseNoMarkup,
};
exports.rawRaw = {
    kind: tag_definitions_1.TagKind.Raw,
    parseLiquidInBody: false,
    parse: parseNoMarkup,
};
exports.javascriptRaw = {
    kind: tag_definitions_1.TagKind.Raw,
    parseLiquidInBody: true,
    parse: parseNoMarkup,
};
exports.schemaRaw = {
    kind: tag_definitions_1.TagKind.Raw,
    parseLiquidInBody: false,
    parse: parseNoMarkup,
};
exports.styleRaw = {
    kind: tag_definitions_1.TagKind.Raw,
    parseLiquidInBody: true,
    parse: parseNoMarkup,
};
exports.stylesheetRaw = {
    kind: tag_definitions_1.TagKind.Raw,
    parseLiquidInBody: true,
    parse: parseNoMarkup,
};
