// @shopify/liquid-html-parser/dist/tags/render.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.includeTag = exports.renderTag = void 0;
const tokenizer_1 = require("../markup/tokenizer");
const types_1 = require("../types");
const tag_definitions_1 = require("../tag-definitions");
// Looks ahead for a render/include named-argument key at the current cursor.
// Ruby render/include strict2_parse keys are a BARE id immediately followed by a
// colon (`key = p.consume; p.consume(:colon)`) — NOT a dotted path. So we match
// only `Id Colon` here. A dotted key like `key.secondkey:` does not look like a
// named argument and is left unconsumed, which then fails the end-of-string check
// below (mirroring Ruby's `p.consume(:colon)` failure on `key.secondkey`).
function looksLikeNamedArgument(markup) {
    return markup.look(tokenizer_1.MarkupTokenType.Id) && markup.look(tokenizer_1.MarkupTokenType.Colon, 1);
}
function parseRenderMarkup(_name, markup, _parser, allowAnyTemplateName = false) {
    const snippet = markup.valueExpression();
    // Ruby `render` requires a quoted string template name at parse time
    // (render.rb `strict2_template_name`: `p.consume(:string)`), whereas `include`
    // accepts any expression and only raises "Illegal template name" at render
    // time when it does not evaluate to a String (include.rb:40). So for `include`
    // we accept any value expression here and defer the type check to the
    // render-tree handler; `render` keeps the stricter parse-time restriction.
    if (!allowAnyTemplateName &&
        snippet.type !== types_1.NodeTypes.String &&
        snippet.type !== types_1.NodeTypes.VariableLookup) {
        throw new Error(`Expected string or variable lookup for snippet, got ${snippet.type}`);
    }
    let variable = null;
    let alias = null;
    let kind = null;
    const kwStart = markup.peek().start;
    if (markup.id('for')) {
        kind = 'for';
    }
    else if (markup.id('with')) {
        kind = 'with';
    }
    if (kind !== null) {
        const name = markup.valueExpression();
        variable = {
            type: types_1.NodeTypes.RenderVariableExpression,
            kind,
            name,
            position: { start: kwStart, end: name.position.end },
            source: name.source,
        };
    }
    // Handle 'as alias' — may appear after 'with expr' / 'for expr' or standalone
    const asStart = markup.peek().start;
    if (markup.id('as')) {
        const aliasToken = markup.consume(tokenizer_1.MarkupTokenType.Id);
        alias = {
            type: types_1.NodeTypes.RenderAliasExpression,
            value: aliasToken.value,
            position: { start: asStart, end: aliasToken.end },
            source: snippet.source,
        };
    }
    const args = [];
    // Commas between named args are optional in Ruby `render`/`include`
    // (render.rb / include.rb strict2_parse: `p.consume?(:comma)` after each
    // attribute). Both `{% render 'x', a: 1, b: 2 %}` and the space-separated
    // `{% render 'x' a: 1 b: 2 %}` are valid. This is a render/include-local loop
    // (NOT a change to the shared `namedArguments()`), so the strict AST for every
    // other tag — and theme-check-common's parse-error detection — is unaffected.
    markup.consumeOptional(tokenizer_1.MarkupTokenType.Comma);
    while (looksLikeNamedArgument(markup)) {
        args.push(markup.namedArgument());
        markup.consumeOptional(tokenizer_1.MarkupTokenType.Comma);
    }
    const end = markup.peek().start;
    return {
        type: types_1.NodeTypes.RenderMarkup,
        snippet,
        variable,
        alias,
        args,
        position: { start: snippet.position.start, end },
        source: snippet.source,
    };
}
exports.renderTag = {
    kind: tag_definitions_1.TagKind.Tag,
    parse: parseRenderMarkup,
};
exports.includeTag = {
    kind: tag_definitions_1.TagKind.Tag,
    parse: (name, markup, parser) => parseRenderMarkup(name, markup, parser, true),
};
